---
title: "Diet_Results_postSRS"
output: html_document
date: "2023-09-26"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Important: 
** The data is already normalized via SRS (scaling with ranked subsampling)



### Prepare work space and load packages

```{r setup, include=FALSE}

##############################################################
library(rlang)
library(cowplot)
library(tidyverse) #
library(readxl) #
library(iNEXT) # 
library(vegan) #
library(reshape2) 
library(ggpubr) # 
library(cowplot) # 
library(devtools) #
library(lme4) 
library(phyloseq)
library(plyr)
library(Rcpp)
library(vctrs)
library(metagMisc)
library(DHARMa)
library(sjPlot)
library(effects)
library(qiime2R)
library(BiocManager)
library(patchwork)
library(microViz)
library(beepr)
library(microViz)
library(mixOmics)
library(miaTime)
library(miaViz)
library(ade4)
library(nlme)
library(pals)
library(RColorBrewer)
library(MetBrewer)
library(MASS)
library(knitr)
library(kableExtra)
library(colorBlindness)
library(devtools) 
library(speedyseq)
library(microbiome)
library(microbiomeutilities)
library(qiime2R)
library(netresponse)

##############################################################

getwd()
# "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/Documents/1. PhD_Main/GitHub_link/Nittedal/Diet_Myotis_brandts_mystacinus"

output <- "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs"

file.name <- "Nittedal_modeling pt1"

todays_date <- Sys.Date()
 
dir.name <- str_c(output,"/", file.name, "_", todays_date)
dir.name
 
output_today <- dir.name
output_today

dir.create(output_today)
output_today

```


## Import and filter the dataset 

- Remove taxa unknow at the Order level and above (none removed)
- Remove OTUs with 10 or less sequence reads (729 OTUs removed) * 
- Remove samples with 600 or less sequence reads (none removed)

```{r}

phylo <- qza_to_phyloseq(
  features= "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/FromFranz_07.2023/R-scrips/sPLS-DA/bat-feat-tab-clean.qza",  
  # This file was sent by Franz on 06.09.2023        
   #it makes sense to use a filtered table, as sPLS-DA can't handle rare features very well
  taxonomy= "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/FromFranz_07.2023/R-scrips/sPLS-DA/bat-classified-repr-seq.qza", 
  metadata= "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/FromFranz_07.2023/R-scrips/sPLS-DA/bat-metadata_clean.tsv")

phylo #shows what our physeq object contains
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 3629 taxa and 64 samples ]
# sample_data() Sample Data:       [ 64 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 3629 taxa by 7 taxonomic ranks ]



sort(colSums(otu_table(phylo))) # 4000 reads per sample 

#we remove all features that are unassigned on the phylum level (none lost)
phylos <- subset_taxa(phylo, Phylum != "NA")

phylos   #shows what our physeq object contains
#phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 3629 taxa and 64 samples ]
# sample_data() Sample Data:       [ 64 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 3629 taxa by 7 taxonomic ranks ]
#now we extract the data we need in the exact format that we need


#we remove all features that are unassigned on Order level (none lost)
phyloseq2 <- subset_taxa(phylo, Order != "NA")

phyloseq2   #shows what our physeq object contains
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 3629 taxa and 64 samples ]
# sample_data() Sample Data:       [ 64 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 3629 taxa by 7 taxonomic ranks ]*


batsbugs <- phylo %>%  ps_filter(bat.sp %in% c("MBRA", "MMYS")) 
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 2616 taxa and 47 samples ]
# sample_data() Sample Data:       [ 47 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 2616 taxa by 7 taxonomic ranks ]
#   
#Inspect the Phyloseq object:
  
  sample_names(batsbugs)
  rank_names(batsbugs)
  sample_variables(batsbugs)
  
sort(colSums(otu_table(batsbugs)))
quantile(colSums(otu_table(batsbugs)))
#   0%  25%  50%  75% 100% 
# 4000 4000 4000 4000 4000 


#Remove OTUs with less than 10 sequence reads 
batsbugs1 <- prune_taxa(taxa_sums(batsbugs) >= 10, batsbugs) 
batsbugs
batsbugs1
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 1887 taxa and 47 samples ]:
# sample_data() Sample Data:        [ 47 samples by 9 sample variables ]:
# tax_table()   Taxonomy Table:     [ 1887 taxa by 7 taxonomic ranks ]:
# taxa are rows

ntaxa(batsbugs1)
# 2616 - 1887 # 729 taxa lost 

quantile(colSums(otu_table(batsbugs1)))
#     0%    25%    50%    75%   100% 
# 3857.0 3934.0 3950.0 3960.5 4000.0 

sort(colSums(otu_table(batsbugs1)))

batsbugs2<-prune_samples(colSums(otu_table(batsbugs1))>=600,batsbugs1) 
batsbugs2 # none lost 
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 1887 taxa and 47 samples ]
# sample_data() Sample Data:       [ 47 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 1887 taxa by 7 taxonomic ranks ]

#removed 5 samples, check that this worked: 
sort(colSums(otu_table(batsbugs2)))
quantile(colSums(otu_table(batsbugs2)))
#     0%    25%    50%    75%   100% 
# 3857.0 3934.0 3950.0 3960.5 4000.0 

## Visualize to check: 
# Compare abundance across of different order of OTUs betwween bat species  
# 
# p1 <- plot_bar(batsbugs2, x = "bat.sp", fill = "Order") +
#                 geom_bar(stat="identity") + 
#   theme_bw() + ggtitle("OTU > 10 reads, Samples > 600 reads")
# p1 # 


```

## Visualize abundance and relative abundance of different prey items 

### Relative abundance representation
- Family 
- Order 

Reference for the code: 
https://mibwurrepo.github.io/Microbial-bioinformatics-introductory-course-Material-2018/composition-plots.html


!! Question !! 
When I plot relative abundance aggregated for taxa Family and compare between the two bats, there apeears to be far fewer reads per sample for MBRA than MMYS but when I compare on the Order level, then this differnce is not so pronounced. Is it because NAs are kicked out? 

```{r}

## Plot relative abundance of different prey taxa at different taxonomic scales


### Family 

# bb2.fam <- aggregate_top_taxa2(batsbugs2, "Family", top = 100)
# 
# plot.composition.CountAbun <- plot_composition(bb2.fam) + 
#   theme_bw() +
#   ggtitle("Top 70 famillies") + 
#   facet_wrap(~ bb2.fam@sam_data$bat.sp) +
#   theme(legend.position = "bottom") 
#   
# plot.composition.CountAbun
# 
# 
# # Abundance 
# 
# bb.fam.rel <- microbiome::transform(bb2.fam, "compositional")
# 
# plot_bar(bb.fam.rel, x="bat.sp", fill = "Family")+ xlab(NULL) + geom_bar(stat="identity") 
# 
# plot_composition(bb.fam.rel, fill = "family", average_by = "bat.sp") + 
#   labs( x = "Bat Species", y = "Relative abundance", title = "Relative Abundance of Prey Item by Family") 
# 
# 
# #### Order
# 
# bb2.order <- aggregate_top_taxa2(batsbugs2, "Order", top = 45) 
# 
# plot.composition.CountAbun <- plot_composition(bb2.order) + 
#   theme_bw() + ggtitle("Top 45 orders") + facet_wrap(~ bb2.order@sam_data$bat.sp) +
#   theme(legend.position = "bottom") 
#   
# plot.composition.CountAbun
# 
# 
# #Make relative abundance 
# 
# 
# bb.order.rel <- transform(bb2.order, "compositional")
# 
# plot_bar(bb.order.rel, x="bat.sp", fill = "Order")+ xlab(NULL) + 
#   geom_bar(stat="identity") + 
#   labs( y = "Relative abundance",
#   title = "Relative Abundance of Prey Item by Order")  +
#   theme(legend.position = "bottom") +
#   theme_bw()
# 
# 
# 
# plot_composition(bb.order.rel, fill = "Order", average_by = "bat.sp") + 
#   labs( x = "Bat Species", y = "Relative abundance", 
#         title = "Relative Abundance of Prey Item by Order") +
#   theme_bw()
# 
# plot.composition.relAbun <- plot_composition(bb.order.rel, sample.sort = "bat.sp") + 
#   facet_wrap(~ bb.order.rel@sam_data$bat.sp) +
#   theme_bw() + theme(text = element_text(size = 20))   
# print(plot.composition.relAbun)


```


## Determine what the NA prey items are 
!! Note !! 
NCBI BLAST of the one genetic sequence that accounts  (one prey item ?) accounting for 411 OTUs .
BLAST result: (Ula species, Crane flies) 
380 of the 411 OTUs were in the Diptera Order:

I assigned 17860 of these reads (380 OTUs) to Ula species but left the rest as unknown orders.
```{r}

############# Find the NAs 

# Convert Phyloseq object to dataframe
bbdf = psmelt(batsbugs2)
dim(bbdf)
#88689    19

summary(bbdf)
names(bbdf)
#  [1] "OTU"            "Sample"         "Abundance"      "sample.nr"      "batch"         
#  [6] "bat.sp"         "sex"            "year"           "date"           "date.text"     
# [11] "nanodrop.ngul." "extradctio.run" "Kingdom"        "Phylum"         "Class"         
# [16] "Order"          "Family"         "Genus"          "Species" 



bbdf1<- bbdf[is.na(bbdf$Family),] # Create a dataset of only prey items with unknown famillies 
summary(bbdf1)
# 19317 obs of NA on at least the family level. 

# But how many unique ASVs? 
bbdf1$OTU <- as.factor(bbdf1$OTU)
summary(bbdf1)


na_asv <- bbdf1 %>% distinct(OTU, .keep_all=T) 
# 411 distinct ASV features that are NA 
# write.csv(na_asv, file=file.path(output_today,"unknown_asvs.csv")) 
# output_today

## Check which sequences are related to these features
seqs <- read_qza("C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Analysis archive/bat_repr-seq.qza")
seqs1 <- seqs$data # extract the data with the DNA sequences 
na_asv$uuid <- na_asv$OTU

na_seq <- cross_join(na_asv, seqs1, copy = TRUE)
dim(na_seq)
# 2065686      21

na_seq1 <- na_seq %>% distinct(OTU, .keep_all = TRUE) 
dim(na_seq)
#write.csv(na_seq1, file=file.path(output_today,"unknown_asvs_with_sequences.csv")) 

na_seq2 <- na_seq1 %>% dplyr::rename(sequences = x) %>% dplyr::select(OTU, sequences)
#write.csv(na_seq2, file=file.path(output_today,"unknown_asvs_with_sequences_only.csv"))

test1 <- unique(na_seq2$sequences)
test1
## There is just one sequence that is related to all of these features... 
# "AGATATTGGAACTATATATAATATAAAAATAGCTTGATCAGGAATAGTGGGGACTTCTTTAAGTATATTAATTCGAGCTGAATTAGGACATCCTGGGGCATTAATTGGAGATGATCAAATTTATAATGTAATTGTTACTGCTCATGCTTTCATTATAATTTTTTTTATAGTAATACCAATTATAATT"

# I manually ran this sequence through blastn on the NCBI Blast+ webpage. 
# The results of the blast are saved :
# C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Analysis archive/BLASToutput_UnknownNA_taxa.txt

# Basically, there were a few different species of Ula (crane flies) that this most likely be. This is the same sequence for all 411 missing features. So now the next step is to insert the family and genus for Ula spp. for all the missing features. 

# Create a list of all the unknown feature IDs
unknowns <- as.character(unique(na_seq2$OTU))
head(unknowns)
unknowns


# Visualize the tax table 
# suppressPackageStartupMessages(library(microViz))
#tax_fix_interactive(batsbugs2)

# Clean the tax names 
batsbugsx <- batsbugs2%>%
 tax_fix(min_length = 0,
  unknowns = c("NA"),
  sep = " ", anon_unique = TRUE,
  suffix_rank = "classified") # Now any unknown family is labeled by its Order 

# Subset to only include the unknown feature IDs, then change unknown Diptera to Ula spp.
my_subset <- subset(otu_table(batsbugsx), rownames(otu_table(batsbugsx)) %in% unknowns)

new_physeq <- merge_phyloseq(my_subset, tax_table(batsbugsx), sample_data(batsbugsx))
new_physeq
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 411 taxa and 47 samples ]:
# sample_data() Sample Data:        [ 47 samples by 9 sample variables ]:
# tax_table()   Taxonomy Table:     [ 411 taxa by 7 taxonomic ranks ]:
# taxa are rows

# How many of these are Diptera? 
dip_new <- subset_taxa(new_physeq, Order == "Diptera")
dip_new 
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 380 taxa and 47 samples ]:
# sample_data() Sample Data:        [ 47 samples by 9 sample variables ]:
# tax_table()   Taxonomy Table:     [ 380 taxa by 7 taxonomic ranks ]:
# taxa are rows

## 380 taxa will be assigned to Ula spp. 

unk.prey <- ps_melt(new_physeq) %>% mutate(Order = factor(Order)) %>% distinct()
summary(unk.prey$Order) 
    # Araneae     Diptera Hymenoptera Lepidoptera  Neuroptera 
    #     376       17860         141         893          47 
# I will be able to assigned 17860 reads of these OTUs to Ula species but will leave the rest as unknown orders. 

new_physeq1 <- new_physeq %>% 
  mutate_tax_table(
  Family = case_when(
  Order  == "Diptera" ~ "Pediciidae", 
  .default = as.character(Family))) %>% 
  mutate_tax_table(Genus = case_when(
  Order  == "Diptera" ~ "Ula",
  .default = as.character(Genus))) %>% 
  mutate_tax_table(Species = case_when(
  Order  == "Diptera" ~ "Ula spp.",
  .default = as.character(Species)))

new_physeq1 # These are the unknowns 
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 411 taxa and 47 samples ]:
# sample_data() Sample Data:        [ 47 samples by 9 sample variables ]:
# tax_table()   Taxonomy Table:     [ 411 taxa by 7 taxonomic ranks ]:
# taxa are rows

# microbiome::plot_taxa_prevalence(new_physeq1, "Family")
# 
# pB <- plot_bar(new_physeq, x = "bat.sp", fill = "Genus") + 
#   xlab(NULL) + geom_bar(stat="identity") + theme_bw() 
# pB
# 
# pB1 <- plot_bar(new_physeq1, x = "bat.sp", fill = "Genus") + 
#   xlab(NULL) + geom_bar(stat="identity") + theme_bw() 
# pB1
# 
# # Large proportion of the unknowns were Ula spp. 
# pB2 <- plot_bar(new_physeq1, x = "bat.sp", fill = "Species") + 
#   xlab(NULL) + geom_bar(stat="identity") + theme_bw() 
# pB2

# what were the OTUs we changed in the end? 

altered <- rownames_to_column(as.data.frame(new_physeq1@tax_table@.Data), "OTU")
altered1 <- altered %>% filter(Family == "Pediciidae") # 380 changed obs! 

nothis <- as.character(unique(altered1$OTU))

# That worked! All the Diptera are now in the family Pediciidae with Genus Ula. 

## Now I just need to merge it back in:


oldtax <- rownames_to_column(as.data.frame(batsbugsx@tax_table@.Data), "OTU") %>% 
  filter(!OTU %in% nothis)  # use the cleaned whole dataset but remove the altered OTUs
#1507

newtax <- rownames_to_column(as.data.frame(new_physeq1@tax_table@.Data), "OTU") %>% filter(Family == "Pediciidae") # Now have only the data of the edited OTUs
# 380 obs

mergetax <- merge(newtax, oldtax, all.x = TRUE, all.y = TRUE) #1887 obs
# 1507+380 = 1887 , good! 

# Now re-insert this data back into the OTU table 

taxmat = as.matrix(mergetax) # also 1887 taxa 
test <- batsbugsx@tax_table # 1887 taxa 

rownames(taxmat) <- taxmat[,1]
taxmat <- taxmat[,-1]
TAX = tax_table(taxmat)


## For reporting in the results section 
batsbugsy <- merge_phyloseq(TAX, sample_data(batsbugsx), otu_table(batsbugsx))
batsbugsy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 1887 taxa and 47 samples ]:
# sample_data() Sample Data:        [ 47 samples by 9 sample variables ]:
# tax_table()   Taxonomy Table:     [ 1887 taxa by 7 taxonomic ranks ]:
# taxa are rows

  sample_names(batsbugsy)
  rank_names(batsbugsy)
  sample_variables(batsbugsy)
  
sort(colSums(otu_table(batsbugsy)))
quantile(colSums(otu_table(batsbugsy)))
summary(colSums(otu_table(batsbugsy)))
   # Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   # 3857    3934    3950    3949    3960    4000 


# Quick check 
plot_bar(batsbugsy, x = "bat.sp", fill = "Order") +
  xlab(NULL) + geom_bar(stat="identity") + theme_bw() + 
  scale_colour_brewer(palette="Set1") + 
  theme(text = element_text(size = 20))  

batsbugsy <- batsbugsy%>%
 tax_fix(min_length = 0,
  unknowns = c("NA"),
  sep = " ", anon_unique = TRUE,
  suffix_rank = "classified") # Now any unknown family is labeled by its Order 
checkthis <- ps_melt(batsbugsy) # That worked 

## Export this phyloseq object so you dont need to recreate it every time!

# saveRDS(batsbugsy, file = file.path(output_today, "BatsBugs_Nasfixed.RDS")) 
# output_today
#testx <- readRDS("C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs/Nittedal_modeling pt1_2023-09-26/BatsBugs_Nasfixed.RDS") # that works!! 


```


###  Calculate prevalence (FOO?) and relative abundance (RRA) 

!! Questions !!: 
I am a little unsure if I am calculating these correctly; would be great if you could double check this section. 

I calculated prevalence and relative abundance for each species separately for visualizing purposes and for the table that is in the manuscript but used the combined data set for ordination. 

When I calculate relative abundance and prevalence using the complete data set with both bat species combined, the prevalence and abundance of different prey taxa is entirely identical between the two bats. But when I calculate this separately then re-combine, I have fewer reads than what I started with. What is the correct approach / what might I be doing wrong? 

```{r,  unaggregated_summaries, results = "asis", echo = FALSE}

######################################################
### All samples combined ### 
######################################################
# test <- prune_taxa(taxa_sums(batsbugsy) > 0, batsbugsy)
# test
# testdf <- ps_melt(test) # still 88689 obs 

# Prevalence of the most abundant ASVs for both bat species - Frequency occurence (FO)? 

### Fora ll samples combined 
prevelancedf = apply(X = otu_table(batsbugsy),
                 MARGIN = 1,
                 FUN = function(x){sum(x > 0)})
# Add taxonomy and total read counts to this data.frame
prevelancedf = data.frame(Prevalence = prevelancedf,
                      TotalAbundance = taxa_sums(batsbugsy),
                      tax_table(batsbugsy))
#write.csv(prevelancedf, file = file.path(output_today, "prevelancedf.csv"))
# output_today
# "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs/Nittedal_modeling pt1_2023-09-26"


## Do this separately for each bat species then recombine with the 
prevall <- prevelancedf %>% rownames_to_column(var = "OTU") %>%
  dplyr::select(OTU, Prevalence, TotalAbundance)

str(prevall)
# 'data.frame':	1887 obs. of  3 variables:
#  $ OTU           : chr  "aedf8e775e4c0c959f9ad1e7460a6b13" "03757d800296133f842997ef7ee29f47" "c36206e2de9ff84834f976b81c27a456" "570987e04aab202503861ef80daf753d" ...
#  $ Prevalence    : int  3 3 3 4 3 2 3 2 8 1 ...
#  $ TotalAbundance: num  992 602 996 970 644 550 426 613 664 206 ...

######################################################

### Calculate relative read abundance
### Not entirely sure how to fit this in just yet. 

## Per sample per prey species 
batsbugysrra  = transform_sample_counts(batsbugsy, function(x) x / sum(x) )
batsbugysrradf <- psmelt(batsbugysrra) # Abundance is now relative abundance 
batsbugsz <- as.data.frame(full_join(batsbugysrradf, prevall, by = "OTU"))
## 88689 obs of 21 variables. 
head(batsbugsz) #
# write.csv(batsbugsz, file = file.path(output_today, "RelativeAbundancePrevalenecCalcAllPreyBothBats.csv"))
# output_today
# "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs/Nittedal_modeling pt1_2023-09-21"


## Looking for top prey items by species instead of OTU - summed the total abundance - not sure if this is a sound approach ** 
topall <- batsbugsz %>% arrange(desc(Prevalence)) %>% 
  group_by(Species) %>% dplyr::summarize(TotalAbundance= sum(TotalAbundance)) %>% 
  mutate(Predator = "Pooled") %>% arrange(desc(TotalAbundance)) %>%  slice(1:10)
kable(topall) %>% kable_minimal()


######################################################
### Myotis brandtii  ### 
######################################################

MBRA <- subset_samples(batsbugsy, bat.sp == "MBRA")
MBRA <- prune_taxa(taxa_sums(MBRA) > 0, MBRA) # Remove taxa not found in MBRA samples 
MBRA
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 774 taxa and 16 samples ]:
# sample_data() Sample Data:        [ 16 samples by 9 sample variables ]:
# tax_table()   Taxonomy Table:     [ 774 taxa by 7 taxonomic ranks ]:
# taxa are rows

### TRY THIS 
MBRA.fam <- aggregate_top_taxa2(MBRA, "Family", top = 70)

plot.composition.CountAbun <- plot_composition(MBRA.fam) + 
  theme_bw() +
  ggtitle("Top 70 famillies") + 
  facet_wrap(~ MBRA.fam@sam_data$bat.sp) +
  theme(legend.position = "bottom") 
  
plot.composition.CountAbun

### Fora ll samples combined 
prevelancedf = apply(X = otu_table(MBRA),
                 MARGIN = 1,
                 FUN = function(x){sum(x > 0)})
# Add taxonomy and total read counts to this data.frame
prevelancedf = data.frame(Prevalence = prevelancedf,
                      TotalAbundance = taxa_sums(MBRA),
                      tax_table(MBRA))
prevalentpreyMBRA <- prevelancedf
# write.csv(prevalentpreyMBRA, file = file.path(output_today, "prevalentpreyMBRA.csv"))
# output_today
# "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs/Nittedal_modeling pt1_2023-09-21"

## Do this separately for each bat species then recombine with the 
prevMBRA <- prevalentpreyMBRA %>% rownames_to_column(var = "OTU") %>% 
  dplyr::select(OTU, Prevalence, TotalAbundance)

str(prevMBRA)

### Calculate relative read abundance

## Per sample per prey species 
batsbugysrraB  = transform_sample_counts(MBRA, function(x) x / sum(x) )
batsbugysrradfB <- psmelt(batsbugysrraB) # abundance is now relative abundance 
batsbugszB <- as.data.frame(full_join(batsbugysrradfB, prevMBRA, by = "OTU"))

head(batsbugszB) 
batsbugszB$BatSpecies <- "M. brandtii" #
# write.csv(batsbugszB, file = file.path(output_today, "RelativeAbundancePrevalenecCalcAllPreyMBRA.csv"))
#"C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs/Nittedal_modeling pt1_2023-09-21

## Looking for top prey items by species instead of OTU - summed the total abundance - not sure if this is a sound approach ** 
topallB <- batsbugszB %>% arrange(desc(Prevalence)) %>% 
  group_by(Species) %>% dplyr::summarize(TotalAbundance= sum(TotalAbundance)) %>% 
  mutate(Predator = "M. brandtii") %>% arrange(desc(TotalAbundance)) %>%  slice(1:10)

TaxOrder <- batsbugszB %>% dplyr::select(Species, Order) %>% distinct()
# 82 unique prey items 
topallB <- left_join(topallB, TaxOrder)
kable(topallB) %>% kable_minimal()


######################################################
### Myotis mystacinus  ### 
######################################################

MMYS <- subset_samples(batsbugsy, bat.sp == "MMYS")
MMYS <- prune_taxa(taxa_sums(MMYS) > 0, MMYS) # Remove taxa not found in MMYS samples 
MMYS

# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 1262 taxa and 31 samples ]:
# sample_data() Sample Data:        [ 31 samples by 9 sample variables ]:
# tax_table()   Taxonomy Table:     [ 1262 taxa by 7 taxonomic ranks ]:

### Fora ll samples combined 
prevelancedf = apply(X = otu_table(MMYS),
                 MARGIN = 1,
                 FUN = function(x){sum(x > 0)})

# Add taxonomy and total read counts to this data.frame
prevelancedf = data.frame(Prevalence = prevelancedf,
                      TotalAbundance = taxa_sums(MMYS),
                      tax_table(MMYS))
prevalentpreyMMYS <- prevelancedf
# write.csv(prevalentpreyMMYS, file = file.path(output_today, "prevalentpreyMMYS.csv"))
# output_today
# "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs/Nittedal_modeling pt1_2023-09-21"

## Do this separately for each bat species then recombine with the 
prevMMYS <- prevalentpreyMMYS %>% rownames_to_column(var = "OTU") %>% 
  dplyr::select(OTU, Prevalence, TotalAbundance)

str(prevMMYS)

### Calculate relative read abundance

## Per sample per prey species 
batsbugysrraM  = transform_sample_counts(MMYS, function(x) x / sum(x) )
batsbugysrradfM <- psmelt(batsbugysrraM) # abundance is now relative abundance 
batsbugszM <- as.data.frame(full_join(batsbugysrradfM, prevMMYS, by = "OTU"))

head(batsbugszM) 
batsbugszM$BatSpecies <- "M. mystacinus" #
# write.csv(batsbugszM, file = file.path(output_today, "RelativeAbundancePrevalenecCalcAllPreyMMYS.csv"))
#"C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs/Nittedal_modeling pt1_2023-09-21

## Looking for top prey items by species instead of OTU - summed the total abundance - not sure if this is a sound approach ** 
topallM <- batsbugszM %>% arrange(desc(Prevalence)) %>% 
  group_by(Species) %>% dplyr::summarize(TotalAbundance= sum(TotalAbundance)) %>% 
  mutate(Predator = "M. mystacinus") %>% arrange(desc(TotalAbundance)) %>%  slice(1:10)

TaxOrder <- batsbugszM %>% dplyr::select(Species, Order) %>% distinct()
# 124 unique prey items 
topallM <- left_join(topallM, TaxOrder)
kable(topallM) %>% kable_minimal()


######################################################

## Combine the two - for visualizing differences 

batsbugszBM <- full_join(batsbugszM, batsbugszB)
summary(batsbugszBM) # 51506 rows 

# write.csv(batsbugszBM, file = file.path(output_today, "bats_abundance_prevalence_Separate_then_combined.csv")) 
# output_today
# "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs/Nittedal_modeling pt1_2023-09-26/bats_abundance_prevalence_Separate_then_combined.csv"
# 12384 + 39122 # good 
mdf <- ps_melt(MMYS)
bdf <- ps_melt(MBRA)
# 39122 + 12384 (dims for MMYS and MBRA) = 551506 (This makes sense...)
batsbugsyy <- prune_taxa()
bbdf <- ps_melt(batsbugsy)
# 

MMYS
MBRA
batsbugsy

# Visualize and compare - ORders 
# pB <- plot_bar(MBRA, x = "bat.sp", fill = "Order") + 
#   xlab(NULL) + geom_bar(stat="identity") + theme_bw()     
# #pB
# 
# 
# pM <- plot_bar(MMYS, x = "bat.sp", fill = "Order") +
#   xlab(NULL) + geom_bar(stat="identity") + theme_bw() 
# #pM
# 
# cowplot::plot_grid(pB, pM) + 
#   scale_colour_brewer(palette="Set1") + 
#   theme(text = element_text(size = 20)) 



```


## ALTERNATIVE 
- With this approach the process of calcultaing relative abundance is a bit more streamlined but there are repeated taxa in some cases because multiple ASVs/OTUs are assocaited with the same taxa...
```{r}

## MBRA

# Agglomerate taxa at genus level
MBRA.spec <- tax_glom(MBRA, taxrank = "Species")

# Top N taxa
N <- 10
top <- names(sort(taxa_sums(MBRA.spec), decreasing = TRUE))[1:N]

# Calculate relative abundance
MBRA.spec.prop <- transform_sample_counts(MBRA.spec, function(x) x / sum(x) )

# Subset object to top N taxa
MBRA.spec.prop.top <- prune_taxa(top, MBRA.spec.prop)

topall1 <- MBRA.spec.prop.top %>% psmelt() %>% head(10)

topall1 %>%  dplyr::select(Order, Species) %>% kable() %>% kable_minimal()

## MMYS

# Agglomerate taxa at genus level
MMYS.spec <- tax_glom(MMYS, taxrank = "Species")

# Top N taxa
N <- 10
top <- names(sort(taxa_sums(MMYS.spec), decreasing = TRUE))[1:N]

# Calculate relative abundance
MMYS.spec.prop <- transform_sample_counts(MMYS.spec, function(x) x / sum(x) )

# Subset object to top N taxa
MMYS.spec.prop.top <- prune_taxa(top, MMYS.spec.prop)

topall1 <- MMYS.spec.prop.top %>% psmelt() %>% head(10)

topall1 %>%  dplyr::select(Species, Order, Abundance) %>% kable() %>% kable_minimal()

```


## Plotting abundance, prevalence, relative abundance across prey orders 

!!Problem!!
When I compare the abundance and prevalence of prey items between bat species I get completely different results depending on whether or not this was calculated for each species separately or if it was calculated in the same data set... Is that to be expected? Have I calculated prevalence and abundance correctly? 

```{r}
prey1 <- batsbugszBM # 51506 obs - prevalence and abundance calculated separate for each bat species then recombined into one dataset 
prey2 <- batsbugsz # 88689 obs - prevalence and abundance calculated in the complete dataset. 

## Just differences in number of counts 
  ggplot() + 
  geom_count(data = prey1, aes(x = BatSpecies, y = Family, color = Order), alpha = 0.5) + 
  scale_size(range = c(5, 15)) +
  theme_bw() + xlab("") + ylab("") + 
  theme(text = element_text(size = 20)) # clearly different between bat species 
  
  ggplot() + 
  geom_count(data = prey2, aes(x = bat.sp, y = Family, color = Order), alpha = 0.5) + 
  scale_size(range = c(5, 15)) +
  theme_bw() + xlab("") + ylab("") + 
  theme(text = element_text(size = 20)) ## nearly identical

# Total Abundance 
  ggplot() + 
  geom_count(data = prey1, aes(x = BatSpecies, y = Family, size = TotalAbundance, color = Order), alpha = 0.5) + 
  scale_size(range = c(5, 12)) +
  theme_bw() + xlab("") + ylab("") +
  theme(text = element_text(size = 20)) +
  geom_jitter()

    ggplot() + 
  geom_count(data = prey2, aes(x = bat.sp, y = Family, size = TotalAbundance, color = Order), alpha = 0.5) + 
  scale_size(range = c(5, 12)) +
  theme_bw() + xlab("") + ylab("") +
  theme(text = element_text(size = 20)) +
  geom_jitter()
  
  # Prevalence
  ggplot() + 
  geom_count(data = prey1, aes(x = BatSpecies, y = Family, size = Prevalence, color = Order), alpha = 0.5) + 
  scale_size(range = c(5, 12)) +
  theme_bw() + xlab("") + ylab("") +
  theme(text = element_text(size = 20)) +
  geom_jitter()
  
    ggplot() + 
  geom_count(data = prey2, aes(x = bat.sp, y = Family, size = Prevalence, color = Order), alpha = 0.5) + 
  scale_size(range = c(5, 12)) +
  theme_bw() + xlab("") + ylab("") +
  theme(text = element_text(size = 20)) +
  geom_jitter()
  
################################################# 
### Use this ###
################################################# 
  
  ## FamilyOrders
  checkthis1 <- prey1 %>% dplyr::select(Family, Order) %>% 
    distinct() %>% 
    mutate(Family = factor(Family)) %>% 
    mutate(Order = factor(Order))
  # 53 families, 10 orders 
    checkthis2 <- prey2 %>% dplyr::select(Family, Order) %>% 
    distinct() %>% 
    mutate(Family = factor(Family)) %>% 
    mutate(Order = factor(Order))
  # 53 families, 10 orders  

  ## Aggregate to Family level then use RRA - 
prey3 <- prey1 %>% group_by(BatSpecies, Sample, Family, Order) %>% 
    dplyr::summarize(FamAbun = sum(TotalAbundance)) ## 1909 obs

checkthis <- prey3 %>% dplyr::select(Family, Order) %>% distinct() %>% arrange(Order) 
listcheck <- as.character(checkthis$Family) %>% unique() # list of 53, good 
prey3$Family <- factor(prey3$Family, levels = listcheck) 
prey3$Order <- factor(prey3$Order)

famlist <- unique(prey3$Family)

  
summary(prey3)
  
 #    BatSpecies           Sample                     Family             Order         FamAbun     
 # Length:1909        Length:1909        Limoniidae    :  47   Diptera    :877   Min.   :   18  
 # Class :character   Class :character   Pediciidae    :  47   Lepidoptera:579   1st Qu.:  153  
 # Mode  :character   Mode  :character   Tachinidae    :  47   Trichoptera:141   Median :  617  
 #                                       Tipulidae     :  47   Neuroptera : 78   Mean   : 3160  
 #                                       Chironomidae  :  47   Hymenoptera: 63   3rd Qu.: 2799  
 #                                       Bolitophilidae:  47   Psocodea   : 62   Max.   :50868  
 #                                       (Other)       :1627   (Other)    :109                 
 #  
  
  
## THIS goes into the paper   
# Now the families are arranged by Order!! 
library(MetBrewer)
p <-  ggplot() + 
  geom_count(data = prey3, aes(x = BatSpecies, y = Family, size = FamAbun, color = Order), alpha = 0.9,) + 
  scale_size_continuous(range = c(5, 12)) +
  scale_color_manual(values = met.brewer("Signac", n = 10)) +
  xlab("") + ylab("") +
  theme(
  panel.background = element_rect(fill = "white",
                                colour = "white",
                                size = 0.5, linetype = "solid"),
  panel.grid.major = element_line(size = 0.5, linetype = 'solid',
                                colour = "#7F7F7F"), 
  panel.grid.minor = element_line(size = 0.25, linetype = 'solid',
                                colour = "#7F7F7F")) + 
  theme(text = element_text(size = 20)) +
  geom_jitter()
p + theme(legend.title = element_blank()) 

## here it is just presence and absence 

p <-  ggplot() + 
  geom_point(data = prey3, aes(x = BatSpecies, y = Family, color = Order), size = 6) + 
  scale_color_manual(values = met.brewer("Signac", n = 10)) +
  xlab("") + ylab("") +
  theme(
  panel.background = element_rect(fill = "white",
                                colour = "white",
                                size = 0.5, linetype = "solid"),
  panel.grid.major = element_line(size = 0.5, linetype = 'solid',
                                colour = "#7F7F7F"), 
  panel.grid.minor = element_line(size = 0.25, linetype = 'solid',
                                colour = "#7F7F7F")) + 
  theme(text = element_text(size = 20)) +
  geom_jitter()
p + theme(legend.title = element_blank()) 

# Range of dot sizes between 10000, 20000 and 30000 

###### Proportion of which prey is represented in between the two specie - Relative Abundance
##### No difference if this is relative or total abundance or even prevalence because it is a proportion 

propbar <- ggplot(prey1) +
  geom_bar(aes(x=BatSpecies, y = Prevalence, fill = Order), stat = "identity", position = "fill") + 
  scale_y_continuous(labels = scales::percent) +
  scale_fill_manual(values = met.brewer("Signac", n = 10)) +
  xlab("") + ylab("") +
  theme(
  panel.background = element_rect(fill = "white",
                                colour = "white",
                                size = 0.5, linetype = "solid"),
  panel.grid.major = element_line(size = 0.5, linetype = 'solid',
                                colour = "#7F7F7F"), 
  panel.grid.minor = element_line(size = 0.25, linetype = 'solid',
                                colour = "#7F7F7F")) + 
  theme(text = element_text(size = 25)) 
propbar +
  theme(legend.position = "bottom") 


### Simple bipartite

bipart <- ggplot(prey1, aes(x = BatSpecies, y = Order, fill = TotalAbundance)) +
  geom_tile() +
  scale_fill_gradient(low = "gray", high = "black") + 
  labs(title = "",
       x = "",
       y = "") +  
  theme(text = element_text(size = 25)) +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  theme(legend.position = "none") 
bipart


```


## Quick summaries for the results section 

```{r}
# Summaries for the first paragraph of the results section 
bbdf <- psmelt(batsbugsy) # 88689 obs of 19 vars


bbdf <- bbdf %>% mutate(across( c(Sample, bat.sp, Order, Family, Genus, Species), factor)) 
str(bbdf)
# 'data.frame':	88689 obs. of  19 variables:
#  $ OTU           : chr  "a7e16d1e1095aa617d8e54d2c373869e" "69f761880a791a38df0fd6953cca5bfa" "0e4ac3759486bca145e8be797c7fc542" "fe67553c5b555066baf8153308ebf221" ...
#  $ Sample        : Factor w/ 47 levels "10B","11B","12B",..: 17 17 32 34 32 32 28 24 16 1 ...
#  $ Abundance     : num  1301 800 699 646 572 ...
#  $ sample.nr     : int  38 38 54 57 54 54 49 44 37 10 ...
#  $ batch         : chr  "B" "B" "A" "B" ...
#  $ bat.sp        : Factor w/ 2 levels "MBRA","MMYS": 1 1 2 2 2 2 2 1 1 2 ...
#  $ sex           : chr  "F" "F" "F" "M" ...
#  $ year          : int  2018 2018 2018 2018 2018 2018 2018 2018 2018 2017 ...
#  $ date          : chr  "6/13/2018" "6/13/2018" "6/18/2018" "6/18/2018" ...
#  $ date.text     : chr  "13_Jun" "13_Jun" "18_Jun" "18_Jun" ...
#  $ nanodrop.ngul.: num  4.9 4.9 5.1 5.1 5.1 5.1 5.7 7.9 6.4 5.9 ...
#  $ extradctio.run: chr  "Ex6" "Ex6" "Ex7" "Ex5" ...
#  $ Kingdom       : chr  "tax=Animalia" "tax=Animalia" "tax=Animalia" "tax=Animalia" ...
#  $ Phylum        : chr  "Arthropoda" "Arthropoda" "Arthropoda" "Arthropoda" ...
#  $ Class         : chr  "Insecta" "Insecta" "Insecta" "Insecta" ...
#  $ Order         : Factor w/ 10 levels "Araneae","Coleoptera",..: 7 6 3 6 3 3 3 6 6 3 ...
#  $ Family        : Factor w/ 53 levels "Alucitidae","Anisopodidae",..: 23 5 31 5 31 31 39 32 5 39 ...
#  $ Genus         : Factor w/ 117 levels "Acompsia","Actia",..: 47 12 95 12 95 95 113 90 12 113 ...
#  $ Species       : Factor w/ 162 levels "Acompsia cinerella",..: 69 18 131 18 131 131 158 123 18 158 ...

# How many prey species for MBRA and MMYS? 

## MBRA
batsbugszB <- batsbugszB %>% mutate(across( c(Sample, bat.sp, BatSpecies, Order, Family, Genus, Species), factor)) 
str(batsbugszB)
# 'data.frame':	12384 obs. of  22 variables:
#  $ OTU           : chr  "a7e16d1e1095aa617d8e54d2c373869e" "69f761880a791a38df0fd6953cca5bfa" "9d507c657353e7e5ae2f41b42b1895ef" "75c9eb1e7b3ce0fd1de14477210a362a" ...
#  $ Sample        : Factor w/ 16 levels "1B","2A","33B",..: 7 7 14 6 7 6 2 10 13 13 ...
#  $ Abundance     : num  0.325 0.2 0.13 0.129 0.119 ...
#  $ sample.nr     : int  38 38 44 37 38 37 2 40 43 43 ...
#  $ batch         : chr  "B" "B" "B" "B" ...
#  $ bat.sp        : Factor w/ 1 level "MBRA": 1 1 1 1 1 1 1 1 1 1 ...
#  $ sex           : chr  "F" "F" "F" "F" ...
#  $ year          : int  2018 2018 2018 2018 2018 2018 2017 2018 2018 2018 ...
#  $ date          : chr  "6/13/2018" "6/13/2018" "6/14/2018" "6/13/2018" ...
#  $ date.text     : chr  "13_Jun" "13_Jun" "14_Jun" "13_Jun" ...
#  $ nanodrop.ngul.: num  4.9 4.9 7.9 6.4 4.9 6.4 4.7 4.9 6.8 6.8 ...
#  $ extradctio.run: chr  "Ex6" "Ex6" "Ex4" "Ex7" ...
#  $ Kingdom       : chr  "tax=Animalia" "tax=Animalia" "tax=Animalia" "tax=Animalia" ...
#  $ Phylum        : chr  "Arthropoda" "Arthropoda" "Arthropoda" "Arthropoda" ...
#  $ Class         : chr  "Insecta" "Insecta" "Insecta" "Insecta" ...
#  $ Order         : Factor w/ 6 levels "Diptera","Hemiptera",..: 5 4 4 4 4 4 1 4 4 4 ...
#  $ Family        : Factor w/ 36 levels "Argyresthiidae",..: 14 1 21 1 1 1 10 36 36 36 ...
#  $ Genus         : Factor w/ 62 levels "Aedes","Ancylis",..: 22 4 45 4 4 4 23 20 20 20 ...
#  $ Species       : Factor w/ 82 levels "Aedes communis",..: 30 6 54 6 6 6 31 26 26 26 ...
#  $ Prevalence    : int  3 1 1 1 2 1 1 1 1 1 ...
#  $ TotalAbundance: num  1395 800 512 508 552 ...
#  $ BatSpecies    : Factor w/ 1 level "M. brandtii": 1 1 1 1 1 1 1 1 1 1 ...

## MMYS 
batsbugszM <- batsbugszM %>% mutate(across( c(Sample, bat.sp, BatSpecies, Order, Family, Genus, Species), factor)) 
str(batsbugszM)
 # $ Sample        : Factor w/ 31 levels "10B","11B","12B",..: 16 18 16 16 13 1 4 2 12 16 ...
 # $ Abundance     : num  0.175 0.163 0.143 0.131 0.13 ...
 # $ sample.nr     : int  54 57 54 54 49 10 13 11 48 54 ...
 # $ batch         : chr  "A" "B" "A" "A" ...
 # $ bat.sp        : Factor w/ 1 level "MMYS": 1 1 1 1 1 1 1 1 1 1 ...
 # $ sex           : chr  "F" "M" "F" "F" ...
 # $ year          : int  2018 2018 2018 2018 2018 2017 2017 2017 2018 2018 ...
 # $ date          : chr  "6/18/2018" "6/18/2018" "6/18/2018" "6/18/2018" ...
 # $ date.text     : chr  "18_Jun" "18_Jun" "18_Jun" "18_Jun" ...
 # $ nanodrop.ngul.: num  5.1 5.1 5.1 5.1 5.7 5.9 5.8 4.9 4.9 5.1 ...
 # $ extradctio.run: chr  "Ex7" "Ex5" "Ex7" "Ex7" ...
 # $ Kingdom       : chr  "tax=Animalia" "tax=Animalia" "tax=Animalia" "tax=Animalia" ...
 # $ Phylum        : chr  "Arthropoda" "Arthropoda" "Arthropoda" "Arthropoda" ...
 # $ Class         : chr  "Insecta" "Insecta" "Insecta" "Insecta" ...
 # $ Order         : Factor w/ 9 levels "Araneae","Coleoptera",..: 3 5 3 3 3 3 3 3 3 3 ...
 # $ Family        : Factor w/ 43 levels "Alucitidae","Anisopodidae",..: 25 5 25 25 31 31 31 38 31 25 ...
 # $ Genus         : Factor w/ 95 levels "Acompsia","Actia",..: 75 11 75 75 92 92 92 83 92 75 ...
 # $ Species       : Factor w/ 124 levels "Acompsia cinerella",..: 99 15 99 99 121 121 121 108 119 99 ...
 # $ Prevalence    : int  1 1 1 3 1 1 1 1 1 1 ...
 # $ TotalAbundance: num  699 646 572 883 519 507 498 499 486 480 ...
 # $ BatSpecies    : Factor w/ 1 level "M. mystacinus": 1 1 1 1 1 1 1 1 1 1 ...


## top taxa



```

## Iris plots 
```{r}


## Same palette as was used in the earlier figure 
## Double check that this is color blind dsafe  
met.brewer("Signac", n = 10)

## What is already in the bar plot
cvdPlot(replacePlotColor(
  displayColors(c(
    "#D8443C", "#9F5691", "#633372", 
    "#E87E8B", "#92C051", "#1F6E9C",
    "#F4C40F", "#DE597C", "#2B9B81", "#FE9B00")))) 

  displayColors(c(
    "#D8443C", "#9F5691", "#633372", 
    # Diptera   #Leptdoptera #Neuroptera
    "#E87E8B", "#92C051", "#1F6E9C",
    #Hymenoptera #Trichoptera #Orthoptera 
    "#F4C40F", "#DE597C", "#2B9B81", "#FE9B00"))
    #Araneae   #Hemioptera  #Psocodea #Coleoptera 
  
  ################################################

  
batsbugsy %>% 
  tax_transform(trans = "clr", rank = "Order") %>% 
  ord_calc() %>% 
  ord_plot_iris(
    tax_level = "Order", 
    ord_plot = "none", 
    anno_colour = "bat.sp",
    anno_binary_style = list(y = 1.05, colour = "gray50", shape = "circle", size = 2)) + 
  scale_fill_manual(values = c(
    "#D8443C", "#9F5691", "#633372", 
    "#E87E8B", "#92C051", "#1F6E9C",
    "#F4C40F", "#DE597C", "#2B9B81", "#FE9B00")) +
    scale_color_manual(values = c("black", "gray")) +
  theme(legend.position = "bottom")


## Separate iris plots for each bat species, side by side: 
MBRA %>% 
  tax_transform(trans = "clr", rank = "Order") %>% 
  ord_calc() %>% 
  ord_plot_iris(tax_level = "Order", ord_plot = "none", anno_colour = "bat.sp") +
    scale_fill_manual(values = c(    
    "#D8443C", "#9F5691", "#633372", 
    "#E87E8B", "#DE597C", "#92C051")) +
  scale_color_manual(values = "white") +
  theme(legend.position = "none") +
MMYS %>% 
  tax_transform(trans = "clr", rank = "Order") %>% 
  ord_calc() %>% 
  ord_plot_iris(tax_level = "Order", ord_plot = "none", anno_colour = "bat.sp") +
    scale_fill_manual(values = c(    
    "#D8443C", "#9F5691", "#633372", 
    "#1F6E9C", "#F4C40F", "#92C051", 
    "#2B9B81", "#E87E8B", "#FE9B00")) +
   theme(legend.position = "none") +
    scale_color_manual(values = "white")  

# checked with legends that the colors matched and then removed for the final figure. 
test <- tax_fix(batsbugsy)

### Composition bar plots
test %>%  
  comp_barplot(tax_level = "Order") +
  scale_fill_manual(values = c(
    "#D8443C", "#9F5691", "#633372", 
    "#E87E8B", "#92C051", "#1F6E9C",
    "#F4C40F", "#DE597C", "#2B9B81", "#FE9B00")) +
    scale_color_manual(values = "white") +
  coord_flip() +
  theme(text = element_text(size = 25))+
  theme(legend.position = "none")


MBRA %>%
  comp_barplot(tax_level = "Order") +
  scale_fill_manual(values = c(    
    "#D8443C", "#9F5691", "#633372", 
    "#E87E8B", "#DE597C", "#92C051")) +
    scale_color_manual(values = "white") +
  coord_flip() +
  theme(text = element_text(size = 25)) +
  theme(legend.position = "none")

MMYS %>%
  comp_barplot(tax_level = "Order") +
  scale_fill_manual(values = c(    
    "#D8443C", "#9F5691", "#633372", 
    "#1F6E9C", "#F4C40F", "#92C051", 
    "#2B9B81", "#DE597C", "#FE9B00")) +
    scale_color_manual(values = "white") +
  coord_flip() +
  theme(text = element_text(size = 25))  +
  theme(legend.position = "none")

#¤ Just for the legend 
batsbugsy %>%
  comp_barplot(tax_level = "Order") +
  scale_fill_manual(labels = c("Diptera", "Lepidoptera", "Neuroptera",
                               "Hymenoptera", "Trichoptera", "Orthoptera",
                               "Araneae", "Hemioptera", "Psocodea"), 
                    values = c(    
    "#D8443C", "#9F5691", "#633372", 
    "#E87E8B", "#92C051", "#1F6E9C",
    "#F4C40F", "#DE597C", "#2B9B81", "#FE9B00")) +
    scale_color_manual(values = "white") +
  coord_flip() +
  theme(text = element_text(size = 25))+
  theme(legend.position = "bottom")

```

## Shannon and Simpson diversity 

Shannon: How difficult it is to predict the identity of a randomly chosen individual.
Simpson: The probability that two randomly chosen individuals are the same species
Fisher: This measures the evenness with which individuals are divided among the taxa present

```{r}
sh <- estimate_richness(batsbugsy, split = TRUE, measures = "Shannon") %>% rownames_to_column("Sample")


# Diversity within each sample 
bbdf_sh <- left_join(bbdf, sh)
bbdf_sh %>% ggplot() + 
  geom_bar(aes(x = Sample, y = Shannon, color = bat.sp), stat = "identity")

plot_richness(batsbugsy) 

# M <- plot_richness(MMYS)
# M
# 
# B <- plot_richness(MBRA)
# B



# For each sample 
batsbugsy@sam_data$sampleid <- as.factor(batsbugsy@sam_data$sample.nr)
MBRA@sam_data$sampleid <- as.factor(MBRA@sam_data$sample.nr)
MMYS@sam_data$sampleid <- as.factor(MMYS@sam_data$sample.nr)

plot_richness(batsbugsy, x="sampleid", measures="Shannon")+
  geom_boxplot(aes(fill = bat.sp), color = "gray") +
  scale_colour_manual(values = batcolors,  aesthetics = c("fill", "colour"), 
                      name = "Bat species") +
  facet_wrap(~bat.sp) +
  theme_bw() # Useful to look at but not something I feel needs to be in the manuscript

# shannonB <- plot_richness(MBRA, x="sampleid", measures="Shannon")+
#   geom_boxplot(fill = "#b3cde0", color = "gray", width = 2) +
#   theme_bw() + 
#   ggtitle("M. brandtii") + 
#   theme(text = element_text(size = 15)) + xlab("")
# shannonB
# 
# shannonM <- plot_richness(MMYS, x="sampleid", measures="Shannon")+
#   geom_boxplot(fill = "#011f4b", color = "gray") +
#   theme_bw() + 
#   ggtitle("M. mystacinus") +
#   theme(text = element_text(size = 15)) + xlab("")
# shannonM


## Alpha diversity - diversity between communities 
rich <- estimate_richness(batsbugsy)
head(rich)
#      Observed Chao1  se.chao1      ACE    se.ACE  Shannon   Simpson InvSimpson    Fisher
# X10B       33    33 0.0000000      NaN       NaN 3.202442 0.9484766   19.40864  4.930898
# X11B       37    37 0.0000000 37.00000 0.9863939 3.146580 0.9421180   17.27653  5.642515
# X12B       65    65 0.1240347 65.67607 2.5055369 3.665782 0.9647341   28.35602 11.034699
# X13B       45    45 0.0000000      NaN       NaN 3.260423 0.9449219   18.15604  7.120491
# X14B       43    43 0.0000000 43.00000 0.9883037 3.404255 0.9568050   23.15084  6.752266
# X15B       58    58 0.4956709 59.08000 1.6013106 3.708866 0.9691568   32.42208  9.645112

# pairwise.wilcox.test(rich$Observed, sample_data(batsbugsy)$bat.sp)
# # 	Pairwise comparisons using Wilcoxon rank sum test with continuity correction 
# # 
# # data:  rich$Observed and sample_data(batsbugsy)$bat.sp 
# # 
# #      MBRA
# # MMYS 0.37
# # 
# # P value adjustment method: holm 
# # Warning message:
# # In wilcox.test.default(xi, xj, paired = paired, ...) :
# #   cannot compute exact p-value with ties
# 
# pairwise.wilcox.test(rich$Shannon, sample_data(batsbugsy)$bat.sp)
# 
# # 	Pairwise comparisons using Wilcoxon rank sum exact test 
# # 
# # data:  rich$Shannon and sample_data(batsbugsy)$bat.sp 
# # 
# #      MBRA
# # MMYS 0.34
# # 
# # P value adjustment method: holm 
# 
# pairwise.wilcox.test(rich$Simpson, sample_data(batsbugsy)$bat.sp)
# # 	Pairwise comparisons using Wilcoxon rank sum exact test 
# # 
# # data:  rich$Simpson and sample_data(batsbugsy)$bat.sp 
# # 
# #      MBRA
# # MMYS 0.4 
# # 
# # P value adjustment method: holm
# 
# pairwise.wilcox.test(rich$Fisher, sample_data(batsbugsy)$bat.sp)
# # 	Pairwise comparisons using Wilcoxon rank sum exact test 
# 
# data:  rich$Simpson and sample_data(batsbugsy)$bat.sp 
# 
#      MBRA
# MMYS 0.4 
# 
# P value adjustment method: holm

## t-test and ANOVA

rich1 <- rich %>% rownames_to_column("Sample") 
rich1$Sample <- gsub("X", "", rich1$Sample)

samplemap <- batsbugsz %>% dplyr::select(Sample, bat.sp, sex, year, date) %>% distinct() #metadata for each sample
rich2 <- left_join(rich1, samplemap) # Each samples metadata and alpha diversity measurements 

hist(rich2$Shannon, main="Shannon index", xlab="") 
# this is not normally distributed so actually no AOV or Tukey tests should not be applied. 
# Instead I can run a glmm to test

library(nlme)
m1 <- lme(Shannon ~ bat.sp*year, random=~1|Sample, data=rich2)
summary(m1) #
# Linear mixed-effects model fit by REML
#   Data: rich2 
#        AIC      BIC    logLik
#   53.65551 64.22271 -20.82775
# 
# Random effects:
#  Formula: ~1 | Sample
#         (Intercept)  Residual
# StdDev:   0.3300256 0.1238386
# 
# Fixed effects:  Shannon ~ bat.sp * year 
#                     Value Std.Error DF    t-value p-value
# (Intercept)      -3.80903  410.6388 43 -0.0092759  0.9926
# bat.spMMYS      193.23219  487.2329 43  0.3965910  0.6936
# year              0.00361    0.2035 43  0.0177207  0.9859
# bat.spMMYS:year  -0.09578    0.2415 43 -0.3966577  0.6936
#  Correlation: 
#                 (Intr) bt.MMYS year  
# bat.spMMYS      -0.843               
# year            -1.000  0.843        
# bat.spMMYS:year  0.843 -1.000  -0.843
# 
# Standardized Within-Group Residuals:
#         Min          Q1         Med          Q3         Max 
# -1.41804764 -0.17744410  0.04025917  0.24637826  0.40965665 
# 
# Number of Observations: 47
# Number of Groups: 47 

### Model predicitons 

interpol <- expand.grid(year = (unique(m1$data$year)),
                        bat.sp = (unique(m1$data$bat.sp)),
                        Sample = unique(m1$data$Sample))

interpol$Shannon<- predict(m1, interpol, type= "response",
                       allow.new.levels= T)


## Plot the predictions and the raw data 
batcolors = c("#b3cde0", "#011f4b") 

predictplot <- ggplot(interpol) +
  geom_boxplot(aes(x = bat.sp, y= Shannon, fill = bat.sp), color = c("black", "gray")) + 
  theme_bw() +
  scale_colour_manual(values = batcolors,  aesthetics = c("fill", "colour"), 
                      name = "Bat species") +
  scale_x_discrete(labels=c("M. brandtii", "M. mystacinus")) +
  xlab("") +
  theme(text = element_text(size = 15)) +
  theme(legend.position="none", 
        axis.text.x=element_text(angle=45,hjust=1,vjust=1,size=12)) + 
  ggtitle("Predicted data") + ylab("Alpha diversity (Shannon)")
predictplot # And now they look the same:( 



## Shannon diversity compared between the two bat species 
rawplot <- plot_richness(batsbugsy, x="bat.sp", measures="Shannon")+
  geom_boxplot(aes(fill = bat.sp), color = c("black", "gray")) + 
  theme_bw() +
  scale_colour_manual(values = batcolors,  aesthetics = c("fill", "colour"), 
                      name = "Bat species") +
  scale_x_discrete(labels=c("M. brandtii", "M. mystacinus")) +
  xlab("") + ylab("Alpha diversity (Shannon)") +
  theme(strip.background = element_blank(), strip.text.x = element_text(size=0)) +
  theme(text = element_text(size = 15)) +
  theme(legend.position="none", 
        axis.text.x=element_text(angle=45,hjust=1,vjust=1,size=12)) +
  ggtitle("Raw data")

rawplot

cowplot::plot_grid(rawplot, predictplot)


```




## Different ordination approach 
https://david-barnett.github.io/microViz/articles/web-only/ordination.html

```{r}

pca <- batsbugsy %>% 
   tax_transform(trans = "log", rank = "Family") %>% 
   dist_calc(dist = "bray") %>% 
   ord_calc("PCoA") %>% ord_plot(color = "bat.sp", size = 2) # can get pca scores 

pca1 <- ordinate(batsbugsy, method = "PCoA", distance = "bray", trans = "log", k = 2)
summary(pca1)

## For plotting

batcolors = c("#b3cde0", "#011f4b") 
          #     MBRA        MMYS 

### Pcoa - Bray - Curtis, log10 transformed - use this figure for the manuscript 
batsbugsy %>%
  tax_transform("log", rank = "Family") %>%
  dist_calc(dist = "bray") %>% #aitchison = euclidian in clr transformed data 
  ord_calc("PCoA") %>%
  ord_plot(color = "bat.sp", shape = "year", size = 4, alpha = 0.8) +
  scale_colour_manual(values = batcolors,  aesthetics = c("fill", "colour"), 
                      name = "Bat species") +
  stat_ellipse(geom = "polygon", type="t", alpha=0.3, aes(fill =bat.sp)) +
  theme_bw() +
  ggside::geom_xsidedensity(aes(fill = bat.sp), alpha = 0.5, show.legend = FALSE) +
  ggside::geom_ysidedensity(aes(fill = bat.sp), alpha = 0.5, show.legend = FALSE) +
  ggside::theme_ggside_void() + 
  theme(legend.position = "none")


### Pcoa - Bray - Curtis, not transformed 
batsbugsy %>%
  tax_transform("identity", rank = "Family") %>%
  dist_calc(dist = "bray") %>% #aitchison = euclidian in clr transformed data 
  ord_calc("PCoA") %>%
  ord_plot(color = "bat.sp", shape = "year", size = 4, alpha = 0.8) +
  scale_colour_manual(values = batcolors,  aesthetics = c("fill", "colour"), 
                      name = "Bat species") +
  stat_ellipse(geom = "polygon", type="t", alpha=0.3, aes(fill =bat.sp)) +
  theme_bw() +
  ggside::geom_xsidedensity(aes(fill = bat.sp), alpha = 0.5, show.legend = FALSE) +
  ggside::geom_ysidedensity(aes(fill = bat.sp), alpha = 0.5, show.legend = FALSE) +
  ggside::theme_ggside_void() + 
  theme(legend.position = "none")

## 
batsbugsy %>%
  tax_transform("log10", rank = "Order") %>%
  dist_calc(dist = "bray") %>% #aitchison = euclidian in clr transformed data 
  ord_calc("PCoA") %>%
  ord_plot(color = "bat.sp", shape = "year", size = 4, alpha = 0.8) +
  scale_colour_manual(values = batcolors,  aesthetics = c("fill", "colour"), 
                      name = "Bat species") +
  stat_ellipse(geom = "polygon", type="t", alpha=0.3, aes(fill =bat.sp)) +
  theme_bw() +
  ggside::geom_xsidedensity(aes(fill = bat.sp), alpha = 0.5, show.legend = FALSE) +
  ggside::geom_ysidedensity(aes(fill = bat.sp), alpha = 0.5, show.legend = FALSE) +
  ggside::theme_ggside_void() + 
  theme(legend.position = "bottom")

#Warning message:
# In microbiome::transform(otu, transform = trans, target = "OTU",  :
#   OTU table contains zeroes. Using log10(1 + x) transform.

### Pcoa - Bray - curtis, relative abundance 
batsbugsy %>%
  tax_transform("compositional", rank = "Family") %>%
  dist_calc(dist = "bray") %>% 
  ord_calc("PCoA") %>%
  ord_plot(color = "bat.sp", shape = "year", size = 4, alpha = 0.8) +
  scale_colour_manual(values = batcolors,  aesthetics = c("fill", "colour"), 
                      name = "Bat species") +
  stat_ellipse(geom = "polygon", type="t", alpha=0.3, aes(fill =bat.sp)) +
  theme_bw() +
  ggside::geom_xsidedensity(aes(fill = bat.sp), alpha = 0.5, show.legend = FALSE) +
  ggside::geom_ysidedensity(aes(fill = bat.sp), alpha = 0.5, show.legend = FALSE) +
  ggside::theme_ggside_void() + 
  theme(legend.position = "none")

###########################################
## combine pca scores with batsbugs object
###########################################

pcadf <- pca$data

bbdf1 <- left_join(bbdf, pcadf)
head(bbdf1)


```


# Mixed linear model 

https://support.minitab.com/en-us/minitab/18/help-and-how-to/modeling-statistics/anova/how-to/mixed-effects-model/interpret-the-results/key-results/#step-2-determine-whether-the-fixed-effect-terms-significantly-affect-the-response

```{r, echo = TRUE ,results = 'hide',fig.keep = 'all'}

### use NMDS1 as the response variable 
library(nlme)

# simplest model 
mi <- lme(MDS1 ~ bat.sp, random=~1|sampleid, data=bbdf1)
summary(mi) # species significant
# Linear mixed-effects model fit by REML
#   Data: bbdf1 
#        AIC      BIC  logLik
#   -5333342 -5333305 2666675
# 
# Random effects:
#  Formula: ~1 | sampleid
#         (Intercept)     Residual
# StdDev:   0.7916966 2.076935e-14
# 
# Fixed effects:  MDS1 ~ bat.sp 
#                  Value   Std.Error    DF   t-value p-value
# (Intercept)  0.3284397 0.008340768 88642  39.37763       0
# bat.spMMYS  -0.7752206 0.010270088    45 -75.48335       0
#  Correlation: 
#            (Intr)
# bat.spMMYS -0.812
# 
# Standardized Within-Group Residuals:
#         Min          Q1         Med          Q3         Max 
# -2.61928949 -0.81251429  0.01737284  0.49178496  2.61928949 
# 
# Number of Observations: 88689
# Number of Groups: 47



# + year
mii <- lme(MDS1~bat.sp + year, random=~1|sampleid, data=bbdf1)
summary(mii) # species and year significant 
# Linear mixed-effects model fit by REML
#   Data: bbdf1 
#        AIC      BIC  logLik
#   -5336306 -5336259 2668158
# 
# Random effects:
#  Formula: ~1 | sampleid
#         (Intercept)     Residual
# StdDev:   0.8085271 2.042184e-14
# 
# Fixed effects:  MDS1 ~ bat.sp + year 
#                  Value   Std.Error    DF   t-value p-value
# (Intercept)  0.5465725 0.009105451 88642  60.02696       0
# bat.spMMYS  -0.8523200 0.010195013    44 -83.60165       0
# year        -0.0000879 0.000001595    44 -55.14862       0
#  Correlation: 
#            (Intr) b.MMYS
# bat.spMMYS -0.784       
# year       -0.434  0.137
# 
# Standardized Within-Group Residuals:
#         Min          Q1         Med          Q3         Max 
# -2.51163939 -0.73392060 -0.05844183  0.64965564  2.81608053 
# 
# Number of Observations: 88689
# Number of Groups: 47

# species*year interaction 
miii <- lme(MDS1~bat.sp + bat.sp*year, random=~1|sampleid, data=bbdf1)
summary(miii)

# Linear mixed-effects model fit by REML
#   Data: bbdf1 
#        AIC      BIC  logLik
#   -5340926 -5340870 2670469
# 
# Random effects:
#  Formula: ~1 | sampleid
#         (Intercept)    Residual
# StdDev:   0.8090302 1.98937e-14
# 
# Fixed effects:  MDS1 ~ bat.sp + bat.sp * year 
#                      Value   Std.Error    DF    t-value p-value
# (Intercept)      0.9632098 0.010727067 88642   89.79247       0
# bat.spMMYS      -1.3789148 0.012520505    43 -110.13252       0
# year            -0.0002559 0.000002886    43  -88.67867       0
# bat.spMMYS:year  0.0002365 0.000003425    43   69.06416       0
#  Correlation: 
#                 (Intr) bt.MMYS year  
# bat.spMMYS      -0.857               
# year            -0.667  0.572        
# bat.spMMYS:year  0.562 -0.609  -0.843
# 
# Standardized Within-Group Residuals:
#        Min         Q1        Med         Q3        Max 
# -2.7122580 -0.8761821  0.1046396  0.5497066  2.7680658 
# 
# Number of Observations: 88689
# Number of Groups: 47 

miv <- lme(MDS1~bat.sp + bat.sp*year + 
           bat.sp*Order, 
           random=~1|sampleid, data=bbdf1)
summary(miv) ## massive ouput, here is just the top part 

# Linear mixed-effects model fit by REML
#   Data: bbdf1 
#        AIC      BIC  logLik
#   -5339681 -5339456 2669865
# 
# Random effects:
#  Formula: ~1 | sampleid
#         (Intercept)     Residual
# StdDev:   0.8090342 1.989566e-14
# 
# Fixed effects:  MDS1 ~ bat.sp + bat.sp * year + bat.sp * Order 
#                                  Value   Std.Error    DF    t-value p-value
# (Intercept)                  0.9632125 0.010728424 88624   89.78136  0.0000
# bat.spMMYS                  -1.3789047 0.012522160    43 -110.11716  0.0000 *
# year                        -0.0002559 0.000002887    43  -88.65406  0.0000 *
# OrderColeoptera              0.0000000 0.000000000 88624    0.00000  1.0000
# OrderDiptera                 0.0000000 0.000000000 88624   -0.01357  0.9892
# OrderHemiptera               0.0000000 0.000000000 88624    0.00000  1.0000
# OrderHymenoptera             0.0000000 0.000000000 88624    0.00000  1.0000
# OrderLepidoptera             0.0000000 0.000000000 88624    0.00181  0.9986
# OrderNeuroptera              0.0000000 0.000000000 88624    0.17185  0.8636
# OrderOrthoptera              0.0000000 0.000000000 88624    0.00000  1.0000
# OrderPsocodea                0.0000000 0.000000000 88624    0.00000  1.0000
# OrderTrichoptera             0.0000000 0.000000000 88624    0.00000  1.0000
# bat.spMMYS:year              0.0002365 0.000003425    43   69.04427  0.0000 *



mv <- lme(MDS1~ bat.sp + 
            bat.sp*year + 
            year + Order +
           bat.sp*Order, 
           random=~1|sampleid, data=bbdf1)

summary(mv)
# Linear mixed-effects model fit by REML
#   Data: bbdf1 
#        AIC      BIC  logLik
#   -5339681 -5339456 2669865
# 
# Random effects:
#  Formula: ~1 | sampleid
#         (Intercept)     Residual
# StdDev:   0.8090342 1.989566e-14
# 
# Fixed effects:  MDS1 ~ bat.sp + bat.sp * year + year + Order + bat.sp * Order 
#                                  Value   Std.Error    DF    t-value p-value
# (Intercept)                  0.9632125 0.010728424 88624   89.78136  0.0000
# bat.spMMYS                  -1.3789047 0.012522160    43 -110.11716  0.0000
# year                        -0.0002559 0.000002887    43  -88.65406  0.0000
# OrderColeoptera              0.0000000 0.000000000 88624    0.00000  1.0000
# OrderDiptera                 0.0000000 0.000000000 88624   -0.01357  0.9892
# OrderHemiptera               0.0000000 0.000000000 88624    0.00000  1.0000
# OrderHymenoptera             0.0000000 0.000000000 88624    0.00000  1.0000
# OrderLepidoptera             0.0000000 0.000000000 88624    0.00181  0.9986
# OrderNeuroptera              0.0000000 0.000000000 88624    0.17185  0.8636
# OrderOrthoptera              0.0000000 0.000000000 88624    0.00000  1.0000
# OrderPsocodea                0.0000000 0.000000000 88624    0.00000  1.0000
# OrderTrichoptera             0.0000000 0.000000000 88624    0.00000  1.0000
# bat.spMMYS:year              0.0002365 0.000003425    43   69.04427  0.0000
# bat.spMMYS:OrderColeoptera   0.0000000 0.000000000 88624    0.00000  1.0000
# bat.spMMYS:OrderDiptera      0.0000000 0.000000000 88624    0.02493  0.9801
# bat.spMMYS:OrderHemiptera    0.0000000 0.000000000 88624    0.00000  1.0000
# bat.spMMYS:OrderHymenoptera  0.0000000 0.000000000 88624    0.00000  1.0000
# bat.spMMYS:OrderLepidoptera  0.0000000 0.000000000 88624   -0.02263  0.9819
# bat.spMMYS:OrderNeuroptera   0.0000000 0.000000000 88624   -0.15283  0.8785
# bat.spMMYS:OrderOrthoptera   0.0000000 0.000000000 88624    0.00000  1.0000
# bat.spMMYS:OrderPsocodea     0.0000000 0.000000000 88624    0.00000  1.0000
# bat.spMMYS:OrderTrichoptera  0.0000000 0.000000000 88624    0.00000  1.0000


mvi<- lme(MDS1~ bat.sp + 
            bat.sp*year + 
            year + Family*bat.sp,
           random=~1|sampleid, data=bbdf1)

summary(mvi) # This most closely resemebles the ordination figure so I will use this for now. 

anova(miV, mV) # ? weird that they are the same. 
#   Model df      AIC      BIC  logLik
# miV     1 24 -5339681 -5339456 2669865
# mV      2 24 -5339681 -5339456 2669865


## Ask Katrine what the best way is to compare model selection for these kinds of NMDS models... 

# anova(m1, m2, m3, m4)
# 
# 
# anova(m1_int, m2_int, m3_int, m4_int)


#Quick visualization of difference between the two species:

plot(allEffects(m1), type = "response")
plot(allEffects(m2), type = "response")
plot(allEffects(m3), type = "response")

plot(allEffects(miv), type = "response")
plot(allEffects(mv), type = "response")

# Very little difference between years 

# https://rdrr.io/cran/sjPlot/f/vignettes/plot_marginal_effects.Rmd

#plotting marginal effects
#plot_model(mod1, type = "pred", terms = "bat.sp")
# plot_model(m3, type = "eff", terms = "bat.sp")

```



### Next steps: 
- Proper diversity index plots / table 
- Pros / cons of PCA, Pcoa and NMDS? 
- Model predictions of NLME model 
- Shannon diversity index, within species and between species 


