
---
title: "Nittedal Diet Analyses"
author: "Reed McKay"
date: "4 9 2021"
output: html_document
---


### Previous versions of the script: 

title: "IBBM Diet Analyses"
author: "Reed McKay"
date: "4 5 2021"
output: html_document


title: "Filter/Normal_NMDS_lmer_batsandbugs"
author: "Reed April McKay"
date: "16 2 2021"
output: html_document


## Important: 
** The data is already normalized via SRS (scaling with ranked subsampling)



### Prepare work space and load packages

```{r setup, include=FALSE}
library(rlang)
library(cowplot)
library(tidyverse) #
library(readxl) #
library(iNEXT) # 
library(vegan) #
library(reshape2) 
library(ggpubr) # 
library(cowplot) # 
library(devtools) #
library(lme4) 
library(phyloseq)
library(plyr)
library(Rcpp)
library(vctrs)
library(metagMisc)
library(DHARMa)
library(sjPlot)
library(effects)
library(qiime2R)
library(BiocManager)
library(patchwork)
library(microViz)
library(beepr)
library(microViz)
library(mixOmics)
library(ade4)
library(nlme)
library(pals)
library(RColorBrewer)

# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")

# The following initializes usage of Bioc devel
# BiocManager::install(version='devel')

#remotes::install_github("mikemc/speedyseq")
library(speedyseq)

library(devtools) # Load the devtools package
# install_github("microbiome/microbiome") # Install the package
# devtools::install_github("microsud/microbiomeutilities")
library(microbiome)
library(microbiomeutilities)
library(MASS)
#devtools::install_github("jbisanz/qiime2R")
library(qiime2R)

# Installing from CRAN
#install.packages("sorvi")
library(sorvi)
# Installing from Github
#install_github("antagomir/netresponse")
library(netresponse)

getwd()
# C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/Documents/1. PhD_Main/GitHub_link/Nittedal/Mbra-vs-Mmys-diet-"

output <- "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs"

file.name <- "Nittedal_modeling pt1"

todays_date <- Sys.Date()
 
dir.name <- str_c(output,"/", file.name, "_", todays_date)
dir.name
 
output_today <- dir.name
output_today

dir.create(output_today)
output_today

```

### Import data, create phyloseq object, convert OTU table to presence/absence (optional)

```{r, echo = TRUE ,results = 'hide',fig.keep = 'all'}

phylo <- qza_to_phyloseq(
  features= "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/FromFranz_07.2023/R-scrips/sPLS-DA/bat-feat-tab-clean.qza",  
  # This file was sent by Franz on 06.09.2023        
   #it makes sense to use a filtered table, as sPLS-DA can't handle rare features very well
  taxonomy= "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/FromFranz_07.2023/R-scrips/sPLS-DA/bat-classified-repr-seq.qza", 
  metadata= "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/FromFranz_07.2023/R-scrips/sPLS-DA/bat-metadata_clean.tsv")

phylo #shows what our physeq object contains
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 3629 taxa and 64 samples ]
# sample_data() Sample Data:       [ 64 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 3629 taxa by 7 taxonomic ranks ]



sort(colSums(otu_table(phylo))) # 4000 reads per sample 

#we remove all features that are unassigned on phylum level
phylos <- subset_taxa(phylo, Phylum != "NA")

phylos   #shows what our physeq object contains
#phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 3629 taxa and 64 samples ]
# sample_data() Sample Data:       [ 64 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 3629 taxa by 7 taxonomic ranks ]
#now we extract the data we need in the exact format that we need

#we remove all features that are unassigned on genus level
phyloseq <- subset_taxa(phylo, Genus != "NA")

phyloseq   #shows what our physeq object contains
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 2536 taxa and 64 samples ]
# sample_data() Sample Data:       [ 64 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 2536 taxa by 7 taxonomic ranks ]

# 1093 taxa lost 


#we remove all features that are unassigned on family level
phyloseq1 <- subset_taxa(phylo, Family != "NA")

phyloseq1   #shows what our physeq object contains
# phyloseq1   #shows what our physeq object contains
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 2665 taxa and 64 samples ]
# sample_data() Sample Data:       [ 64 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 2665 taxa by 7 taxonomic ranks ]
#3629 - 2665 =  964


#we remove all features that are unassigned on genus level
phyloseq2 <- subset_taxa(phylo, Order != "NA")

phyloseq2   #shows what our physeq object contains
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 3629 taxa and 64 samples ]
# sample_data() Sample Data:       [ 64 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 3629 taxa by 7 taxonomic ranks ]*

### Franz has suggested NOT filtering out taxa with NA values at this stage.


batsbugs <- phylo %>%  ps_filter(bat.sp %in% c("MBRA", "MMYS"))
batsbugs  
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 2616 taxa and 47 samples ]
# sample_data() Sample Data:       [ 47 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 2616 taxa by 7 taxonomic ranks ]
#   
#Inspect the Phyloseq object:
  
  sample_names(batsbugs)
  rank_names(batsbugs)
  sample_variables(batsbugs)
  
sort(colSums(otu_table(batsbugs)))
quantile(colSums(otu_table(batsbugs)))
#   0%  25%  50%  75% 100% 
# 4000 4000 4000 4000 4000 
```


### Filter 
```{r, echo = TRUE ,results = 'hide',fig.keep = 'all'}

#Filter1 = batbugs1 

#Remove OTUs with less than 10 sequence reads 
batsbugs1 <- prune_taxa(taxa_sums(batsbugs) >= 10, batsbugs) 
batsbugs
batsbugs1
# 2616 - 1887 # 729 taxa lost 

ntaxa(batsbugs1)
quantile(colSums(otu_table(batsbugs1)))
#     0%    25%    50%    75%   100% 
# 3857.0 3934.0 3950.0 3960.5 4000.0 

sort(colSums(otu_table(batsbugs1)))

# Remove samples with unusually small amounts of reads  (less tan 600)
# Removed four samples:     13B   3B  49B   5B  11B
# (recommendation from Lisa Lunde)

batsbugs2<-prune_samples(colSums(otu_table(batsbugs1))>=600,batsbugs1) 
batsbugs2 # none lost 
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 1887 taxa and 47 samples ]
# sample_data() Sample Data:       [ 47 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 1887 taxa by 7 taxonomic ranks ]

#removed 5 samples, check that this worked: 
sort(colSums(otu_table(batsbugs2)))
quantile(colSums(otu_table(batsbugs2)))
#     0%    25%    50%    75%   100% 
# 3857.0 3934.0 3950.0 3960.5 4000.0 


# Compare abundance across phyloseq objects 

p0 <- plot_bar(batsbugs, x = "bat.sp", fill = "Order") +
                geom_bar(stat="identity") + theme_bw() + ggtitle("Raw")
p0 # raw data 


p1 <- plot_bar(batsbugs2, x = "bat.sp", fill = "Order") +
                geom_bar(stat="identity") + theme_bw() + ggtitle("Samples > 600 reads")
p1 # batsbugs 2 and removed samples with less than 600 reads 

```

Relative abundance representation
Pooled samples
Normalized ad filtered data 
Top Orders 

https://mibwurrepo.github.io/Microbial-bioinformatics-introductory-course-Material-2018/composition-plots.html

```{r}

## MICROBIOMEUTILITIES IS THE NEW ONE TO USE! 

bb2.fam <- aggregate_top_taxa2(batsbugs2, "Family", top = 70)

plot.composition.CountAbun <- plot_composition(bb2.fam) + 
  theme_bw() +ggtitle("Top 70 famillies")
  
plot.composition.CountAbun


#Make relative abundance 

bb.fam.rel <- microbiome::transform(bb2.fam, "compositional")

plot_bar(bb.fam.rel, x="bat.sp", fill = "Family")+ xlab(NULL) + geom_bar(stat="identity") 

plot_composition(bb.fam.rel, fill = "family", average_by = "bat.sp") + 
  labs( x = "Bat Species", y = "Relative abundance", title = "Relative Abundance of Prey Item by Family") 


```

Now for order

```{r}

bb2.order <- aggregate_top_taxa2(batsbugs2, "Order", top = 45) 

plot.composition.CountAbun <- plot_composition(bb2.order) + 
  theme_bw() + ggtitle("Top 45 orders") + facet_wrap(~ bb2.order@sam_data$bat.sp)
  
plot.composition.CountAbun


#Make relative abundance 


bb.order.rel <- transform(bb2.order, "compositional")

plot_bar(bb.order.rel, x="bat.sp", fill = "Order")+ xlab(NULL) + 
  geom_bar(stat="identity") + 
  labs( y = "Relative abundance",
  title = "Relative Abundance of Prey Item by Order")  +
  theme_bw()



plot_composition(bb.order.rel, fill = "Order", average_by = "bat.sp") + 
  labs( x = "Bat Species", y = "Relative abundance", 
        title = "Relative Abundance of Prey Item by Order") +
  theme_bw()

plot.composition.relAbun <- plot_composition(bb.order.rel, sample.sort = "bat.sp") + 
  facet_wrap(~ bb.order.rel@sam_data$bat.sp) +
  theme_bw() + theme(text = element_text(size = 20))   
print(plot.composition.relAbun)

```


## Finding the unknown prey items associated with NA features. 


```{r, echo = TRUE ,results = 'hide',fig.keep = 'all'}

# Taxa with positive sum across samples
bb_prune <- prune_taxa(taxa_sums(batsbugs2) > 0, batsbugs2)
ntaxa(bb_prune)
# 1887
ntaxa(batsbugs2)
# 1887


############# Find the NAs 

# Convert Phyloseq object to dataframe
bbdf = psmelt(bb_prune)
dim(bbdf)
#88689    19


summary(bbdf)
names(bbdf)
#new_DF <- DF[is.na(DF$Var),]
bbdf1<- bbdf[is.na(bbdf$Family),]
summary(bbdf1)
# 19317 obs of NA on at least the family level. 

# But how many unique ASVs? 
bbdf1$OTU <- as.factor(bbdf1$OTU)
summary(bbdf1)


na_asv <- bbdf1 %>% distinct(OTU, .keep_all=T) 
# 411 distinct ASV features that are NA 
# write.csv(na_asv, file=file.path(output_today,"unknown_asvs.csv")) 
# output_today

## Check which sequences are related to these features
seqs <- read_qza("C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Analysis archive/bat_repr-seq.qza")

seqs1 <- seqs$data # extract the data with the DNA sequences 
na_asv$uuid <- na_asv$OTU

na_seq <- cross_join(na_asv, seqs1, copy = TRUE)
dim(na_seq)
# 2065686      21

na_seq1 <- na_seq %>% distinct(OTU, .keep_all = TRUE) 
dim(na_seq)
#write.csv(na_seq1, file=file.path(output_today,"unknown_asvs_with_sequences.csv")) 

na_seq2 <- na_seq1 %>% dplyr::rename(sequences = x) %>% dplyr::select(OTU, sequences)
#write.csv(na_seq2, file=file.path(output_today,"unknown_asvs_with_sequences_only.csv"))

test1 <- unique(na_seq2$sequences)
test1
## There is just one sequence that is related to all of these features... 
# "AGATATTGGAACTATATATAATATAAAAATAGCTTGATCAGGAATAGTGGGGACTTCTTTAAGTATATTAATTCGAGCTGAATTAGGACATCCTGGGGCATTAATTGGAGATGATCAAATTTATAATGTAATTGTTACTGCTCATGCTTTCATTATAATTTTTTTTATAGTAATACCAATTATAATT"

# I manually ran this sequence through blastn on the NCBI Blast+ webpage. 
# The results of the blast are saved :
# C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Analysis archive/BLASToutput_UnknownNA_taxa.txt

# Basically, there were a few different species of Ula (crane flies) that this most likely be. This is the same sequence for all 411 missing features. So now the next step is to insert the family and genus for Ula spp. for all the missing features. 

# Create a list of all the unknown feature IDs
unknowns <- as.character(unique(na_seq2$OTU))
head(unknowns)
unknowns


# Visualize the tax table 
# suppressPackageStartupMessages(library(microViz))
#tax_fix_interactive(batsbugs2)

# Clean the tax names 
batsbugsx <- batsbugs2%>%
 tax_fix(min_length = 0,
  unknowns = c("NA"),
  sep = " ", anon_unique = TRUE,
  suffix_rank = "classified")

# Subset to only include the unknown feature IDs , then change unknown Diptera to Ula spp.
my_subset <- subset(otu_table(batsbugsx), rownames(otu_table(batsbugsx)) %in% unknowns)

new_physeq <- merge_phyloseq(my_subset, tax_table(batsbugsx), sample_data(batsbugsx))

plot_taxa_prevalence(new_physeq, "Family")

new_physeq1 <- new_physeq %>% 
  mutate_tax_table(
  Family = case_when(
  Order  == "Diptera" ~ "Pediciidae", 
  .default = as.character(Family))) %>% 
  mutate_tax_table(Genus = case_when(
  Order  == "Diptera" ~ "Ula",
  .default = as.character(Genus))) %>% 
  mutate_tax_table(Species = case_when(
  Order  == "Diptera" ~ "Ula spp.",
  .default = as.character(Species)))

new_physeq1
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 411 taxa and 47 samples ]:
# sample_data() Sample Data:        [ 47 samples by 9 sample variables ]:
# tax_table()   Taxonomy Table:     [ 411 taxa by 7 taxonomic ranks ]:
# taxa are rows

plot_taxa_prevalence(new_physeq1, "Family")

pB <- plot_bar(new_physeq, x = "bat.sp", fill = "Genus") + 
  xlab(NULL) + geom_bar(stat="identity") + theme_bw() 
pB

pB1 <- plot_bar(new_physeq1, x = "bat.sp", fill = "Genus") + 
  xlab(NULL) + geom_bar(stat="identity") + theme_bw() 
pB1

pB2 <- plot_bar(new_physeq1, x = "bat.sp", fill = "Species") + 
  xlab(NULL) + geom_bar(stat="identity") + theme_bw() 
pB2

# what were the OTUs we changed in the end? 

altered <- rownames_to_column(as.data.frame(new_physeq1@tax_table@.Data), "OTU")
altered1 <- altered %>% filter(Family == "Pediciidae") # 380 changed obs! 

nothis <- as.character(unique(altered1$OTU))

# That worked! All the Diptera are now in the family Pediciidae with Genus Ula. 

## Now I just need to merge it back in:


oldtax <- rownames_to_column(as.data.frame(batsbugsx@tax_table@.Data), "OTU") %>% 
  filter(!OTU %in% nothis)  # use the cleaned whole dataset but remove the altered OTUs
#1507

newtax <- rownames_to_column(as.data.frame(new_physeq1@tax_table@.Data), "OTU") %>% filter(Family == "Pediciidae") # Now have only the data of the edited OTUs
# 380 obs

mergetax <- merge(newtax, oldtax, all.x = TRUE, all.y = TRUE) #1887 obs
# 1507+380 = 1887 , good! 

# Now re-insert this data back into the OTU table 

taxmat = as.matrix(mergetax) # also 1887 taxa 
test <- batsbugsx@tax_table # 1887 taxa 

rownames(taxmat) <- taxmat[,1]
taxmat <- taxmat[,-1]
TAX = tax_table(taxmat)

batsbugsy <- merge_phyloseq(TAX, sample_data(batsbugsx), otu_table(batsbugsx))
batsbugsy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 1887 taxa and 47 samples ]:
# sample_data() Sample Data:        [ 47 samples by 9 sample variables ]:
# tax_table()   Taxonomy Table:     [ 1887 taxa by 7 taxonomic ranks ]:
# taxa are rows


```

###  Plotting differences between MBRA and MMYS and summarizing top prevalent prey 
```{r}
#Subset the raw data by bat species and summarize OTU abundance

#tax_fix_interactive(batsbugsy)

# MBRA 

MBRA <- subset_samples(batsbugsy, bat.sp == "MBRA")
MBRA


plot_taxa_prevalence(MBRA, "Order") 
quantile(colSums(otu_table(MBRA)))
sort(colSums(otu_table(MBRA)))

######################################################

#MMYS
MMYS <- subset_samples(batsbugsy, bat.sp == "MMYS")
MMYS

plot_taxa_prevalence(MMYS, "Order") 
quantile(colSums(otu_table(MMYS)))
sort(colSums(otu_table(MMYS)))

######################################################


# Visualize and compare - ORders 

pB <- plot_bar(MBRA, x = "bat.sp", fill = "Order") + 
  xlab(NULL) + geom_bar(stat="identity") + theme_bw() + theme(text = element_text(size = 20))       
pB


pM <- plot_bar(MMYS, x = "bat.sp", fill = "Order") +
  xlab(NULL) + geom_bar(stat="identity") + theme_bw() 
pM

cowplot::plot_grid(pB, pM) + 
  scale_colour_brewer(palette="Set1") 

# Visualize and compare - families 
tax_glom(MBRA, "Family") # 53 different families for MBRA
tax_glom(MMYS, "Family") # 53 different families for MMYS

tax_glom(MBRA, "Genus") # 117 different Genera for MBRA
tax_glom(MMYS, "Genus") # 117 different Genera for MMYS



######################################################

# Prevalence of the top 10 most abundant ASVs for both bat species from raw data 
### Now there are some NAs in the top prey items - we may able to figure out what these are later by blasting the data through NCBI

prevelancedf = apply(X = otu_table(batsbugsy),
                 MARGIN = 1,
                 FUN = function(x){sum(x > 0)})
# Add taxonomy and total read counts to this data.frame
prevelancedf = data.frame(Prevalence = prevelancedf,
                      TotalAbundance = taxa_sums(batsbugsy),
                      tax_table(batsbugsy))
prevelancedf[1:20,]


######################################################

# Prevalence of the top 10 most abundant OTUs for MBRA from raw data

prevelancedf = apply(X = otu_table(MBRA),
                 MARGIN = 1,
                 FUN = function(x){sum(x > 0)})
# Add taxonomy and total read counts to this data.frame
prevelancedf = data.frame(Prevalence = prevelancedf,
                      TotalAbundance = taxa_sums(MBRA),
                      tax_table(MBRA))
prevelancedf[1:20,]

######################################################

# Prevalence of the top 10 most abundant OTUs for MMYS from raw data

prevelancedf = apply(X = otu_table(MMYS),
                 MARGIN = 1,
                 FUN = function(x){sum(x > 0)})
# Add taxonomy and total read counts to this data.frame
prevelancedf = data.frame(Prevalence = prevelancedf,
                      TotalAbundance = taxa_sums(MMYS),
                      tax_table(MMYS))
prevelancedf[1:20,]


########################################################

## Export this phyloseq object so you dont need to recreate it every time!

# saveRDS(batsbugsy, file = file.path(output_today, "BatsBugs_Nasfixed.RDS")) 
# output_today
# testx <- readRDS("C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs/Nittedal_modeling pt1_2023-09-18/BatsBugs_Nasfixed.RDS") # that works!! 

```

## Import species information data for the 10 orders of arthropods found in our data that for all the species in those orders documented in Swedens artdatabanken: 


```{r}

# There is one csv file for each Arthropod order, read them all in together

input1 <- "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Inputs/Artfakta.se"

artfakta <- list.files(path=input1, pattern=".csv", recursive = TRUE) # 10 files for for 10 orders -

df <-  lapply(artfakta, function(i){
  read.csv(i, sep = ";")}) 

dfx <- df %>% bind_rows()

#write.csv(dfx, file = file.path(output_today, "BugsCombinedSweden.csv"))
output_today
# [1] "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs/Nittedal_modeling pt1_2023-09-18"


# araneae <- read.csv(artfakta[1], row.names = NULL, sep = ";")
# coleoptera <- read.csv(artfakta[2], row.names = NULL, sep = ";")
# diptera <- read.csv(artfakta[3], row.names = NULL, sep = ";")
# hemiptera <- read.csv(artfakta[4], row.names = NULL, sep = ";")
# hymenoptera <- read.csv(artfakta[5], row.names = NULL, sep = ";") 
# # 1000 ish taxa were cut off - may need to look into this later
# lepidoptera <- read.csv(artfakta[6], row.names = NULL, sep = ";")
# neuoptera <- read.csv(artfakta[7], row.names = NULL, sep = ";")
# orthoptera <- read.csv(artfakta[8], row.names = NULL, sep = ";")
# psocodea <- read.csv(artfakta[9], row.names = NULL, sep = ";")
# trichoptera <- read.csv(artfakta[10], row.names = NULL, sep = ";")



```



 
```{r, echo = TRUE ,results = 'hide',fig.keep = 'all'}



## Remember. This changes each time you run it. Eventually, should just save a csv file of one output for this and reuse it in a separate analysis probably. 
bbord <- ordinate(batsbugsy, "NMDS", distance="bray", k = 5)
bbord
        

# envfit() from vegan to stack on top of ordination. Typically best for continuous variables, also accepts factor variables. This is an alternative to ANOVA. Not a real statistcal test. 

O12<- plot_ordination(batsbugsy, bbord, color = "bat.sp", axes = c(1,2)) + 
  stat_conf_ellipse(aes(fill=bat.sp ),alpha=0.25,geom="polygon",level=0.95) + 
  theme_bw() + 
  ggtitle("NMDS1 and 2 with Bray-Curtis difference index ")
O12


O13<- plot_ordination(batsbugsy, bbord, color = "bat.sp", axes = c(1,3)) + 
  stat_conf_ellipse(aes(fill=bat.sp ),alpha=0.25,geom="polygon",level=0.95) + 
  theme_bw()+ 
  ggtitle("NMDS1 and 3 with Bray-Curtis difference index ")
O13

O14<- plot_ordination(batsbugsy, bbord, color = "bat.sp", axes = c(1,4)) + 
  stat_conf_ellipse(aes(fill=bat.sp ),alpha=0.25,geom="polygon",level=0.95) + 
  theme_bw() + 
  ggtitle("NMDS1 and 4 with Bray-Curtis difference index ")
O14 


O23 <- plot_ordination(batsbugsy, bbord, color = "bat.sp", axes = c(2,3)) + 
  stat_conf_ellipse(aes(fill=bat.sp ),alpha=0.25,geom="polygon",level=0.95) + 
  theme_bw() + 
  ggtitle("NMDS 2 and 3 with Bray-Curtis difference index ")
O23 

O24 <- plot_ordination(batsbugsy, bbord, color = "bat.sp", axes = c(2,4)) + 
  stat_conf_ellipse(aes(fill=bat.sp ),alpha=0.25,geom="polygon",level=0.95) + 
  theme_bw() + 
  ggtitle("NMDS 2 and 4 with Bray-Curtis difference index ")
O24

O34 <- plot_ordination(batsbugsy, bbord, color = "bat.sp", axes = c(3,4)) + 
  stat_conf_ellipse(aes(fill=bat.sp ),alpha=0.25,geom="polygon",level=0.95) + 
  theme_bw() + 
  ggtitle("NMDS 3 and 4 with Bray-Curtis difference index ")
O34


#Create a dataframe with NMDS scores
ordscore <- scores(bbord, choices=c(1,2,3,4,5), tidy = TRUE)
ordscore <- as.data.frame(ordscore)
ordscore <- cbind(rownames(ordscore), data.frame(ordscore, row.names=NULL))
names(ordscore)[1] <- "Sample"
#1934 obs of 8 vars 
 
```

# Mixed linear model 

https://support.minitab.com/en-us/minitab/18/help-and-how-to/modeling-statistics/anova/how-to/mixed-effects-model/interpret-the-results/key-results/#step-2-determine-whether-the-fixed-effect-terms-significantly-affect-the-response


```{r, echo = TRUE ,results = 'hide',fig.keep = 'all'}
bbdf <- ps_melt(batsbugsy) # convert to a df 
bbdf1 <- left_join(bbdf, ordscore) # add the NMDS scores

## use this data frame for modeling further: 

#write.csv(bbdf1, file=file.path(output_today,"mbra_mmys_NMDS_Scores_diet.csv"))

### use NMDS1 as the response variable 
library(nlme)

# simplest model 
m1 <- lme(NMDS1~bat.sp, random=~1|label, data=bbdf1)
summary(m1)

# +year
m2 <- lme(NMDS1~bat.sp + year, random=~1|label,data=bbdf1)
summary(m2)

# species*year interaction 
m3 <- lme(NMDS1~bat.sp + bat.sp*year, random=~1|label, data=bbdf1)
summary(m3)


## Ask Katrine what the best way is to compare model selection for these kinds of NMDS models... 

# anova(m1, m2, m3, m4)
# 
# 
# anova(m1_int, m2_int, m3_int, m4_int)


#Quick visualization of difference between the two species:

par(mfrow=c(2, 1))
plot(allEffects(m1), type = "response")
plot(allEffects(m2), type = "response")
plot(allEffects(m3), type = "response")



# https://rdrr.io/cran/sjPlot/f/vignettes/plot_marginal_effects.Rmd

#plotting marginal effects
#plot_model(mod1, type = "pred", terms = "bat.sp")
plot_model(m3, type = "eff", terms = "bat.sp")

