
---
title: "Nittedal Diet Analyses"
author: "Reed McKay"
date: "4 9 2021"
output: html_document
---


### Previous versions of the script: 

title: "IBBM Diet Analyses"
author: "Reed McKay"
date: "4 5 2021"
output: html_document


title: "Filter/Normal_NMDS_lmer_batsandbugs"
author: "Reed April McKay"
date: "16 2 2021"
output: html_document


## Important: 
** The data is already normalized via SRS (scaling with ranked subsampling)



### Prepare work space and load packages

```{r setup, include=FALSE}
library(rlang)
library(cowplot)
library(tidyverse) #
library(readxl) #
library(iNEXT) # 
library(vegan) #
library(reshape2) 
library(ggpubr) # 
library(cowplot) # 
library(devtools) #
library(lme4) 
library(phyloseq)
library(plyr)
library(Rcpp)
library(vctrs)
library(metagMisc)
library(DHARMa)
library(sjPlot)
library(effects)
library(qiime2R)
library(BiocManager)
library(patchwork)
library(microViz)
library(beepr)
library(microViz)
library(mixOmics)
library(ade4)
library(nlme)
library(pals)
library(RColorBrewer)
library(MetBrewer)
library(MASS)

# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# The following initializes usage of Bioc devel
# BiocManager::install(version='devel')
#remotes::install_github("mikemc/speedyseq")
library(speedyseq)

library(devtools) # Load the devtools package
# install_github("microbiome/microbiome") # Install the package
# devtools::install_github("microsud/microbiomeutilities")
library(microbiome)
library(microbiomeutilities)


#devtools::install_github("jbisanz/qiime2R")
library(qiime2R)

# Installing from CRAN
#install.packages("sorvi")
library(sorvi)

# Installing from Github
#install_github("antagomir/netresponse")
library(netresponse)

getwd()
# C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/Documents/1. PhD_Main/GitHub_link/Nittedal/Mbra-vs-Mmys-diet-"

output <- "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs"

file.name <- "Nittedal_modeling pt1"

todays_date <- Sys.Date()
 
dir.name <- str_c(output,"/", file.name, "_", todays_date)
dir.name
 
output_today <- dir.name
output_today

dir.create(output_today)
output_today

```

### Import data, create phyloseq object, convert OTU table to presence/absence (optional)

```{r, echo = TRUE ,results = 'hide',fig.keep = 'all'}

phylo <- qza_to_phyloseq(
  features= "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/FromFranz_07.2023/R-scrips/sPLS-DA/bat-feat-tab-clean.qza",  
  # This file was sent by Franz on 06.09.2023        
   #it makes sense to use a filtered table, as sPLS-DA can't handle rare features very well
  taxonomy= "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/FromFranz_07.2023/R-scrips/sPLS-DA/bat-classified-repr-seq.qza", 
  metadata= "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/FromFranz_07.2023/R-scrips/sPLS-DA/bat-metadata_clean.tsv")

phylo #shows what our physeq object contains
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 3629 taxa and 64 samples ]
# sample_data() Sample Data:       [ 64 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 3629 taxa by 7 taxonomic ranks ]



sort(colSums(otu_table(phylo))) # 4000 reads per sample 

#we remove all features that are unassigned on phylum level
phylos <- subset_taxa(phylo, Phylum != "NA")

phylos   #shows what our physeq object contains
#phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 3629 taxa and 64 samples ]
# sample_data() Sample Data:       [ 64 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 3629 taxa by 7 taxonomic ranks ]
#now we extract the data we need in the exact format that we need

#we remove all features that are unassigned on genus level
phyloseq <- subset_taxa(phylo, Genus != "NA")

phyloseq   #shows what our physeq object contains
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 2536 taxa and 64 samples ]
# sample_data() Sample Data:       [ 64 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 2536 taxa by 7 taxonomic ranks ]

# 1093 taxa lost 


#we remove all features that are unassigned on family level
phyloseq1 <- subset_taxa(phylo, Family != "NA")

phyloseq1   #shows what our physeq object contains
# phyloseq1   #shows what our physeq object contains
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 2665 taxa and 64 samples ]
# sample_data() Sample Data:       [ 64 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 2665 taxa by 7 taxonomic ranks ]
#3629 - 2665 =  964


#we remove all features that are unassigned on genus level
phyloseq2 <- subset_taxa(phylo, Order != "NA")

phyloseq2   #shows what our physeq object contains
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 3629 taxa and 64 samples ]
# sample_data() Sample Data:       [ 64 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 3629 taxa by 7 taxonomic ranks ]*

### Franz has suggested NOT filtering out taxa with NA values at this stage.


batsbugs <- phylo %>%  ps_filter(bat.sp %in% c("MBRA", "MMYS"))
batsbugs  
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 2616 taxa and 47 samples ]
# sample_data() Sample Data:       [ 47 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 2616 taxa by 7 taxonomic ranks ]
#   
#Inspect the Phyloseq object:
  
  sample_names(batsbugs)
  rank_names(batsbugs)
  sample_variables(batsbugs)
  
sort(colSums(otu_table(batsbugs)))
quantile(colSums(otu_table(batsbugs)))
#   0%  25%  50%  75% 100% 
# 4000 4000 4000 4000 4000 
```


### Filter 
```{r, echo = TRUE ,results = 'hide',fig.keep = 'all'}

#Filter1 = batbugs1 

#Remove OTUs with less than 10 sequence reads 
batsbugs1 <- prune_taxa(taxa_sums(batsbugs) >= 10, batsbugs) 
batsbugs
batsbugs1
# 2616 - 1887 # 729 taxa lost 

ntaxa(batsbugs1)
quantile(colSums(otu_table(batsbugs1)))
#     0%    25%    50%    75%   100% 
# 3857.0 3934.0 3950.0 3960.5 4000.0 

sort(colSums(otu_table(batsbugs1)))

# Remove samples with unusually small amounts of reads  (less tan 600)
# Removed four samples:     13B   3B  49B   5B  11B
# (recommendation from Lisa Lunde)

batsbugs2<-prune_samples(colSums(otu_table(batsbugs1))>=600,batsbugs1) 
batsbugs2 # none lost 
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 1887 taxa and 47 samples ]
# sample_data() Sample Data:       [ 47 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 1887 taxa by 7 taxonomic ranks ]

#removed 5 samples, check that this worked: 
sort(colSums(otu_table(batsbugs2)))
quantile(colSums(otu_table(batsbugs2)))
#     0%    25%    50%    75%   100% 
# 3857.0 3934.0 3950.0 3960.5 4000.0 


# Compare abundance across phyloseq objects 

p0 <- plot_bar(batsbugs, x = "bat.sp", fill = "Order") +
                geom_bar(stat="identity") + theme_bw() + ggtitle("Raw")
p0 # raw data 


p1 <- plot_bar(batsbugs2, x = "bat.sp", fill = "Order") +
                geom_bar(stat="identity") + theme_bw() + ggtitle("Samples > 600 reads")
p1 # batsbugs 2 and removed samples with less than 600 reads 

```

Relative abundance representation
Top families 
Top Orders 

https://mibwurrepo.github.io/Microbial-bioinformatics-introductory-course-Material-2018/composition-plots.html

```{r}

## MICROBIOMEUTILITIES IS THE NEW ONE TO USE! 

bb2.fam <- aggregate_top_taxa2(batsbugs2, "Family", top = 70)

plot.composition.CountAbun <- plot_composition(bb2.fam) + 
  theme_bw() +ggtitle("Top 70 famillies")
  
plot.composition.CountAbun


#Make relative abundance 

bb.fam.rel <- microbiome::transform(bb2.fam, "compositional")

plot_bar(bb.fam.rel, x="bat.sp", fill = "Family")+ xlab(NULL) + geom_bar(stat="identity") 

plot_composition(bb.fam.rel, fill = "family", average_by = "bat.sp") + 
  labs( x = "Bat Species", y = "Relative abundance", title = "Relative Abundance of Prey Item by Family") 


##### Order

bb2.order <- aggregate_top_taxa2(batsbugs2, "Order", top = 45) 

plot.composition.CountAbun <- plot_composition(bb2.order) + 
  theme_bw() + ggtitle("Top 45 orders") + facet_wrap(~ bb2.order@sam_data$bat.sp)
  
plot.composition.CountAbun


#Make relative abundance 


bb.order.rel <- transform(bb2.order, "compositional")

plot_bar(bb.order.rel, x="bat.sp", fill = "Order")+ xlab(NULL) + 
  geom_bar(stat="identity") + 
  labs( y = "Relative abundance",
  title = "Relative Abundance of Prey Item by Order")  +
  theme_bw()



plot_composition(bb.order.rel, fill = "Order", average_by = "bat.sp") + 
  labs( x = "Bat Species", y = "Relative abundance", 
        title = "Relative Abundance of Prey Item by Order") +
  theme_bw()

plot.composition.relAbun <- plot_composition(bb.order.rel, sample.sort = "bat.sp") + 
  facet_wrap(~ bb.order.rel@sam_data$bat.sp) +
  theme_bw() + theme(text = element_text(size = 20))   
print(plot.composition.relAbun)



```


## Finding the unknown prey items associated with NA features. 
```{r, echo = TRUE ,results = 'hide',fig.keep = 'all'}

# Taxa with positive sum across samples
bb_prune <- prune_taxa(taxa_sums(batsbugs2) > 0, batsbugs2)
ntaxa(bb_prune)
# 1887
ntaxa(batsbugs2)
# 1887


############# Find the NAs 

# Convert Phyloseq object to dataframe
bbdf = psmelt(bb_prune)
dim(bbdf)
#88689    19


summary(bbdf)
names(bbdf)
#new_DF <- DF[is.na(DF$Var),]
bbdf1<- bbdf[is.na(bbdf$Family),]
summary(bbdf1)
# 19317 obs of NA on at least the family level. 

# But how many unique ASVs? 
bbdf1$OTU <- as.factor(bbdf1$OTU)
summary(bbdf1)


na_asv <- bbdf1 %>% distinct(OTU, .keep_all=T) 
# 411 distinct ASV features that are NA 
# write.csv(na_asv, file=file.path(output_today,"unknown_asvs.csv")) 
# output_today

## Check which sequences are related to these features
seqs <- read_qza("C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Analysis archive/bat_repr-seq.qza")

seqs1 <- seqs$data # extract the data with the DNA sequences 
na_asv$uuid <- na_asv$OTU

na_seq <- cross_join(na_asv, seqs1, copy = TRUE)
dim(na_seq)
# 2065686      21

na_seq1 <- na_seq %>% distinct(OTU, .keep_all = TRUE) 
dim(na_seq)
#write.csv(na_seq1, file=file.path(output_today,"unknown_asvs_with_sequences.csv")) 

na_seq2 <- na_seq1 %>% dplyr::rename(sequences = x) %>% dplyr::select(OTU, sequences)
#write.csv(na_seq2, file=file.path(output_today,"unknown_asvs_with_sequences_only.csv"))

test1 <- unique(na_seq2$sequences)
test1
## There is just one sequence that is related to all of these features... 
# "AGATATTGGAACTATATATAATATAAAAATAGCTTGATCAGGAATAGTGGGGACTTCTTTAAGTATATTAATTCGAGCTGAATTAGGACATCCTGGGGCATTAATTGGAGATGATCAAATTTATAATGTAATTGTTACTGCTCATGCTTTCATTATAATTTTTTTTATAGTAATACCAATTATAATT"

# I manually ran this sequence through blastn on the NCBI Blast+ webpage. 
# The results of the blast are saved :
# C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Analysis archive/BLASToutput_UnknownNA_taxa.txt

# Basically, there were a few different species of Ula (crane flies) that this most likely be. This is the same sequence for all 411 missing features. So now the next step is to insert the family and genus for Ula spp. for all the missing features. 

# Create a list of all the unknown feature IDs
unknowns <- as.character(unique(na_seq2$OTU))
head(unknowns)
unknowns


# Visualize the tax table 
# suppressPackageStartupMessages(library(microViz))
#tax_fix_interactive(batsbugs2)

# Clean the tax names 
batsbugsx <- batsbugs2%>%
 tax_fix(min_length = 0,
  unknowns = c("NA"),
  sep = " ", anon_unique = TRUE,
  suffix_rank = "classified")

# Subset to only include the unknown feature IDs , then change unknown Diptera to Ula spp.
my_subset <- subset(otu_table(batsbugsx), rownames(otu_table(batsbugsx)) %in% unknowns)

new_physeq <- merge_phyloseq(my_subset, tax_table(batsbugsx), sample_data(batsbugsx))

plot_taxa_prevalence(new_physeq, "Family")

new_physeq1 <- new_physeq %>% 
  mutate_tax_table(
  Family = case_when(
  Order  == "Diptera" ~ "Pediciidae", 
  .default = as.character(Family))) %>% 
  mutate_tax_table(Genus = case_when(
  Order  == "Diptera" ~ "Ula",
  .default = as.character(Genus))) %>% 
  mutate_tax_table(Species = case_when(
  Order  == "Diptera" ~ "Ula spp.",
  .default = as.character(Species)))

new_physeq1 # These are the unknowns 
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 411 taxa and 47 samples ]:
# sample_data() Sample Data:        [ 47 samples by 9 sample variables ]:
# tax_table()   Taxonomy Table:     [ 411 taxa by 7 taxonomic ranks ]:
# taxa are rows

plot_taxa_prevalence(new_physeq1, "Family")

pB <- plot_bar(new_physeq, x = "bat.sp", fill = "Genus") + 
  xlab(NULL) + geom_bar(stat="identity") + theme_bw() 
pB

pB1 <- plot_bar(new_physeq1, x = "bat.sp", fill = "Genus") + 
  xlab(NULL) + geom_bar(stat="identity") + theme_bw() 
pB1

# Large proportion of the unknowns were Ula spp. 
pB2 <- plot_bar(new_physeq1, x = "bat.sp", fill = "Species") + 
  xlab(NULL) + geom_bar(stat="identity") + theme_bw() 
pB2

# what were the OTUs we changed in the end? 

altered <- rownames_to_column(as.data.frame(new_physeq1@tax_table@.Data), "OTU")
altered1 <- altered %>% filter(Family == "Pediciidae") # 380 changed obs! 

nothis <- as.character(unique(altered1$OTU))

# That worked! All the Diptera are now in the family Pediciidae with Genus Ula. 

## Now I just need to merge it back in:


oldtax <- rownames_to_column(as.data.frame(batsbugsx@tax_table@.Data), "OTU") %>% 
  filter(!OTU %in% nothis)  # use the cleaned whole dataset but remove the altered OTUs
#1507

newtax <- rownames_to_column(as.data.frame(new_physeq1@tax_table@.Data), "OTU") %>% filter(Family == "Pediciidae") # Now have only the data of the edited OTUs
# 380 obs

mergetax <- merge(newtax, oldtax, all.x = TRUE, all.y = TRUE) #1887 obs
# 1507+380 = 1887 , good! 

# Now re-insert this data back into the OTU table 

taxmat = as.matrix(mergetax) # also 1887 taxa 
test <- batsbugsx@tax_table # 1887 taxa 

rownames(taxmat) <- taxmat[,1]
taxmat <- taxmat[,-1]
TAX = tax_table(taxmat)

batsbugsy <- merge_phyloseq(TAX, sample_data(batsbugsx), otu_table(batsbugsx))
batsbugsy
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 1887 taxa and 47 samples ]:
# sample_data() Sample Data:        [ 47 samples by 9 sample variables ]:
# tax_table()   Taxonomy Table:     [ 1887 taxa by 7 taxonomic ranks ]:
# taxa are rows

## Export this phyloseq object so you dont need to recreate it every time!

# saveRDS(batsbugsy, file = file.path(output_today, "BatsBugs_Nasfixed.RDS")) 
# output_today
# testx <- readRDS("C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs/Nittedal_modeling pt1_2023-09-18/BatsBugs_Nasfixed.RDS") # that works!! 



```

###  Calculate prevalence (FOO?) and relative abundance (RRA) 
```{r}
######################################################
### All samples combined ### 
######################################################

# Prevalence of the most abundant ASVs for both bat species from raw data 

### Fora ll samples combined 
prevelancedf = apply(X = otu_table(batsbugsy),
                 MARGIN = 1,
                 FUN = function(x){sum(x > 0)})
# Add taxonomy and total read counts to this data.frame
prevelancedf = data.frame(Prevalence = prevelancedf,
                      TotalAbundance = taxa_sums(batsbugsy),
                      tax_table(batsbugsy))
prevalentpreyall <- prevelancedf
#write.csv(prevalentpreyall, file = file.path(output_today, "prevalentpreyall.csv"))
# output_today
# "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs/Nittedal_modeling pt1_2023-09-19"

## Do this separately for each bat species then recombine with the 
prevall <- prevalentpreyall %>% rownames_to_column(var = "OTU") %>% 
  dplyr::select(OTU, Prevalence, TotalAbundance)

str(prevall)
# 'data.frame':	1887 obs. of  3 variables:
#  $ OTU           : chr  "aedf8e775e4c0c959f9ad1e7460a6b13" "03757d800296133f842997ef7ee29f47" "c36206e2de9ff84834f976b81c27a456" "570987e04aab202503861ef80daf753d" ...
#  $ Prevalence    : int  3 3 3 4 3 2 3 2 8 1 ...
#  $ TotalAbundance: num  992 602 996 970 644 550 426 613 664 206 ...

######################################################

### Calculate relative read abundance
### Not entirely sure how to fit this in just yet. 

## Per sample per prey species 
batsbugysrra  = transform_sample_counts(batsbugsy, function(x) x / sum(x) )
batsbugysrradf <- psmelt(batsbugysrra) # abundance is now relative abundance 
batsbugsz <- as.data.frame(full_join(batsbugysrradf, prevall, by = "OTU"))

head(batsbugsz) #
# write.csv(batsbugsz, file = file.path(output_today, "RelativeAbundancePrevalenecCalcAllPreyBothBats.csv"))
# output_today
# "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs/Nittedal_modeling pt1_2023-09-21"

######################################################
### Myotis brandtii  ### 
######################################################

MBRA <- subset_samples(batsbugsy, bat.sp == "MBRA")
MBRA <- prune_taxa(taxa_sums(MBRA) > 0, MBRA) # Remove taxa not found in MBRA samples 
MBRA
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 774 taxa and 16 samples ]:
# sample_data() Sample Data:        [ 16 samples by 9 sample variables ]:
# tax_table()   Taxonomy Table:     [ 774 taxa by 7 taxonomic ranks ]:
# taxa are rows

### Fora ll samples combined 
prevelancedf = apply(X = otu_table(MBRA),
                 MARGIN = 1,
                 FUN = function(x){sum(x > 0)})
# Add taxonomy and total read counts to this data.frame
prevelancedf = data.frame(Prevalence = prevelancedf,
                      TotalAbundance = taxa_sums(MBRA),
                      tax_table(MBRA))
prevalentpreyMBRA <- prevelancedf
# write.csv(prevalentpreyMBRA, file = file.path(output_today, "prevalentpreyMBRA.csv"))
# output_today
# "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs/Nittedal_modeling pt1_2023-09-21"

## Do this separately for each bat species then recombine with the 
prevMBRA <- prevalentpreyMBRA %>% rownames_to_column(var = "OTU") %>% 
  dplyr::select(OTU, Prevalence, TotalAbundance)

str(prevMBRA)

### Calculate relative read abundance

## Per sample per prey species 
batsbugysrraB  = transform_sample_counts(MBRA, function(x) x / sum(x) )
batsbugysrradfB <- psmelt(batsbugysrraB) # abundance is now relative abundance 
batsbugszB <- as.data.frame(full_join(batsbugysrradfB, prevMBRA, by = "OTU"))

head(batsbugszB) 
batsbugszB$BatSpecies <- "M. brandtii" #
# write.csv(batsbugszB, file = file.path(output_today, "RelativeAbundancePrevalenecCalcAllPreyMBRA.csv"))
#"C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs/Nittedal_modeling pt1_2023-09-21


######################################################
### Myotis mystacinus  ### 
######################################################

MMYS <- subset_samples(batsbugsy, bat.sp == "MMYS")
MMYS <- prune_taxa(taxa_sums(MMYS) > 0, MMYS) # Remove taxa not found in MMYS samples 
MMYS
# phyloseq-class experiment-level object
# otu_table()   OTU Table:          [ 1262 taxa and 31 samples ]:
# sample_data() Sample Data:        [ 31 samples by 9 sample variables ]:
# tax_table()   Taxonomy Table:     [ 1262 taxa by 7 taxonomic ranks ]:

### Fora ll samples combined 
prevelancedf = apply(X = otu_table(MMYS),
                 MARGIN = 1,
                 FUN = function(x){sum(x > 0)})

# Add taxonomy and total read counts to this data.frame
prevelancedf = data.frame(Prevalence = prevelancedf,
                      TotalAbundance = taxa_sums(MMYS),
                      tax_table(MMYS))
prevalentpreyMMYS <- prevelancedf
# write.csv(prevalentpreyMMYS, file = file.path(output_today, "prevalentpreyMMYS.csv"))
# output_today
# "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs/Nittedal_modeling pt1_2023-09-21"

## Do this separately for each bat species then recombine with the 
prevMMYS <- prevalentpreyMMYS %>% rownames_to_column(var = "OTU") %>% 
  dplyr::select(OTU, Prevalence, TotalAbundance)

str(prevMMYS)

### Calculate relative read abundance

## Per sample per prey species 
batsbugysrraM  = transform_sample_counts(MMYS, function(x) x / sum(x) )
batsbugysrradfM <- psmelt(batsbugysrraM) # abundance is now relative abundance 
batsbugszM <- as.data.frame(full_join(batsbugysrradfM, prevMMYS, by = "OTU"))

head(batsbugszM) 
batsbugszM$BatSpecies <- "M. mystacinus" #
# write.csv(batsbugszM, file = file.path(output_today, "RelativeAbundancePrevalenecCalcAllPreyMMYS.csv"))
#"C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs/Nittedal_modeling pt1_2023-09-21

######################################################

## Combine the two 

batsbugszBM <- full_join(batsbugszM, batsbugszB)
summary(batsbugszBM) # 51506 rows 
# 12384 + 39122 # good 

# Visualize and compare - ORders 
# pB <- plot_bar(MBRA, x = "bat.sp", fill = "Order") + 
#   xlab(NULL) + geom_bar(stat="identity") + theme_bw()     
# #pB
# 
# 
# pM <- plot_bar(MMYS, x = "bat.sp", fill = "Order") +
#   xlab(NULL) + geom_bar(stat="identity") + theme_bw() 
# #pM
# 
# cowplot::plot_grid(pB, pM) + 
#   scale_colour_brewer(palette="Set1") + 
#   theme(text = element_text(size = 20))   

plot_bar(batsbugsy, x = "bat.sp", fill = "Order") +
  xlab(NULL) + geom_bar(stat="identity") + theme_bw() + 
  scale_colour_brewer(palette="Set1") + 
  theme(text = element_text(size = 20))  


```

## How do the top prey items for MBRA and MMYS relate to their habitat selection? 
- Get arthropod species habitat use data from Artdatabanken.se
- Create a habitat column based on the arthropods 
- Plot the differences in terms of prey abundance and prevalence 

```{r}

# There is one csv file for each Arthropod order, read them all in together
# 
# input1 <- "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Inputs/Artfakta.se"
# 
# 
# artfakta <- list.files(path=input1, pattern=".csv", recursive = TRUE) # 10 files for for 10 orders -
# 
# df <-  lapply(artfakta, function(i){
#   read.csv(i, sep = ";")}) 
# 
# dfx <- df %>% bind_rows()

#write.csv(dfx, file = file.path(output_today, "BugsCombinedSweden.csv"))
# output_today
# [1] "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs/Nittedal_modeling pt1_2023-09-18"


# araneae <- read.csv(artfakta[1], row.names = NULL, sep = ";")
# coleoptera <- read.csv(artfakta[2], row.names = NULL, sep = ";")
# diptera <- read.csv(artfakta[3], row.names = NULL, sep = ";")
# hemiptera <- read.csv(artfakta[4], row.names = NULL, sep = ";")
# hymenoptera <- read.csv(artfakta[5], row.names = NULL, sep = ";") 
# # 1000 ish taxa were cut off - may need to look into this later
# lepidoptera <- read.csv(artfakta[6], row.names = NULL, sep = ";")
# neuoptera <- read.csv(artfakta[7], row.names = NULL, sep = ";")
# orthoptera <- read.csv(artfakta[8], row.names = NULL, sep = ";")
# psocodea <- read.csv(artfakta[9], row.names = NULL, sep = ";")
# trichoptera <- read.csv(artfakta[10], row.names = NULL, sep = ";")


swbugs <- read.csv("C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs/Nittedal_modeling pt1_2023-09-18/BugsCombinedSweden.csv")

## Which of these have a habitat description? 

swbugs1 <- swbugs %>% 
  dplyr::select(Vetenskapligt.namn, Landskapstyp, RedListCategory) %>% 
  dplyr::rename(Species = Vetenskapligt.namn, habitat = Landskapstyp, status = RedListCategory) %>% 
  dplyr::mutate(across(c(Species, status), factor)) 

# swbugsh <- swbugs1 %>% filter(habitat != "") 
# summary(swbugsh) #  15219 obs that have some habitat data 


## Combine with bat data 
prey <- left_join(batsbugszBM, swbugs1, by = "Species", relationship = "many-to-many") ## 63731 obs (now more than batsbugsZBM..) 
prey <- prey %>% mutate(habitat = replace_na(habitat, "Unknown"))

## what kind of habitats are included?
test <- prey %>% mutate(habitat = factor(habitat)) 
levels(test$habitat)

## check if there are any taxa for which both abundance and prevalence are zero 
prey_fix <- subset(prey, prey$Prevalence != 0 & prey$TotalAbundance != 0) # 61407 as well - Nope, good!  


prey0 <- prey_fix %>% filter(habitat == "Unknown") # 21548 obs for which habitat is unknown 

## How much of this is Hymenoptera? 

preyH <- prey0 %>% filter(Order == "Hymenoptera") %>% distinct(Species) ## 3 obs ... too generalize taxa to fix  

## Check to see what taxa this includes 
unknownhabitatprey <- prey0 %>% 
  dplyr::select(Species, Prevalence, Abundance) %>% 
  filter(Prevalence >= 2) %>% # Look for prey items with prevalence higher than 2 
  mutate(Species = factor(Species)) 
manualfix <- levels(unknownhabitatprey$Species) # 9 prey items that match this 
manualfix
# "Helina Genus"             "Pseudatemelia josephinae" "Tipula Genus"             "Tricyphona alticola"      "Ula spp."   

# Export and manually fill in for these species 
#write.csv(manualfix, file = file.path(output_today, "unknownpreytoassign.csv"))
#output_today
# "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs/Nittedal_modeling pt1_2023-09-20"

## I manually searched for these prey items plus some others. 
preyx <- prey %>% 
  dplyr::mutate(habitat1 = case_when(
    Species %in% c("Helina Genus", "Thricops semicinereus") ~ "(S) (L) (V) (J) (U) (H) (F)",
    Species %in% c("Chironomidae Family", "Tricyphona alticola", "Ula mixta", "Ula spp.") ~ "(S) (V) (L) (J)",
    Species %in% c("Tipula Genus", "Tipula iberica", "Tipulidae Family") ~ "(S) (V) (L) (J) (U)",
    Species %in% c("Nemapogon cloacella", "Hemerobius pini", 
                   "Hemerobius perelegans", "Hemerobius fenestratus",
                   "Neolygus contaminatus", "Neuratelia Genus", 
                   "Pseudatemelia josephinae") ~ "(S)", 
    TRUE ~ as.character(habitat)))  


prey1 <- preyx  %>% 
  mutate(Agriculture = case_when(
  str_detect(habitat1, pattern = "(J)") ~ 1, # Agriculture 
  TRUE ~ 0)) %>% 
  mutate(Forest = case_when(
  str_detect(habitat1, pattern = "(S)") ~ 2, # Forest
  TRUE ~ 0)) %>% 
  mutate(Fresh_water = case_when(
  str_detect(habitat1, pattern = "(L)") ~ 3, # Fresh water 
  TRUE ~ 0)) %>% 
  mutate(Wetland = case_when(
  str_detect(habitat1, pattern = "(V)") ~ 4, # Wetland
  TRUE ~ 0)) %>% 
  mutate(Mountain = case_when(
  str_detect(habitat1, pattern = "(F)") ~ 5, # Mountain 
  TRUE ~ 0)) %>% 
  mutate(Urban = case_when(
  str_detect(habitat1, pattern = "(U)") ~ 6, # Urban  
  TRUE ~ 0)) %>%
  mutate(Coastal = case_when(
  str_detect(habitat1, pattern = "(H)") ~ 7, # Coastal  
  TRUE ~ 0)) %>% 
  mutate(Unknown = case_when(
  str_detect(habitat1, pattern = "Unknown") ~ 8,
  TRUE ~ 0)) 

# write.csv(prey1, file = file.path(output_today, "BatPreyDatawHabitat.csv")) 
# "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs/Nittedal_modeling pt1_2023-09-21"

  


##################################################
###### Comparing habitat of prey ######
##################################################

# Prevalence of prey items related back to habitat type - Family 

  checkthis <- prey1 %>% dplyr::select(Family, Order) %>% distinct() %>% arrange(Order) 
  listcheck <- as.character(checkthis$Family) %>% unique() # list of 53, good 
  prey1$Family <- factor(prey1$Family, levels = listcheck) 
  prey1$Order <- factor(prey1$Order) # Create a nice pallete with 10 distinct colors 

p <- ggplot() + 
  geom_count(data = prey1, aes(x = Agriculture, y = Family, size = Prevalence), alpha = 0.5, color = "#ff9700") +
  geom_count(data = prey1, aes(x = Fresh_water, y = Family, size = Prevalence), alpha = 0.5, color = "#0080ff") +
  geom_count(data = prey1, aes(x = Forest, y = Family, size = Prevalence), alpha = 0.5, color = "forestgreen") +
  geom_count(data = prey1, aes(x = Wetland, y = Family, size = Prevalence), alpha = 0.5, color = "#3232a2") + 
  geom_count(data = prey1, aes(x = Mountain, y = Family, size = Prevalence), alpha = 0.5, color = "brown") + 
  geom_count(data = prey1, aes(x = Urban, y = Family, size = Prevalence), alpha = 0.5, color = "#8f6e98") + 
  geom_count(data = prey1, aes(x = Coastal, y = Family, size = Prevalence), alpha = 0.5, color = "salmon") + 
  geom_count(data = prey1, aes(x = Unknown, y = Family, size = Prevalence), alpha = 0.5, color = "black") + 
  facet_wrap(~bat.sp) + 
    theme(
  panel.background = element_rect(fill = "white",
                                colour = "white",
                                size = 0.5, linetype = "solid"),
  panel.grid.major = element_line(size = 0.5, linetype = 'solid',
                                colour = "#7F7F7F"), 
  panel.grid.minor = element_line(size = 0.25, linetype = 'solid',
                                colour = "#7F7F7F")) + 
  xlab("") + ylab("") + 
  scale_size_continuous(range = c(5, 12)) +
  theme(axis.text.x=element_blank(), axis.ticks.x= element_blank()) +
  theme(text = element_text(size = 20)) + xlim(c(1,8)) +
  theme(legend.position = "none") 
  
p # Not possible to see nuanced differences without with Prevalence, but it is more obvious with relative abundance. 

# Prevalence of prey items related back to habitat type - Order 
p <- ggplot() + 
  geom_count(data = prey1, aes(x = Agriculture, y = Order, size = Abundance), alpha = 0.7, color = "#ff9700") +
  geom_count(data = prey1, aes(x = Fresh_water, y = Order, size = Abundance), alpha = 0.7, color = "#0080ff") +
  geom_count(data = prey1, aes(x = Forest, y = Order, size = Abundance), alpha = 0.7, color = "forestgreen") +
  geom_count(data = prey1, aes(x = Wetland, y = Order, size = Abundance), alpha = 0.7, color = "#3232a2") + 
  geom_count(data = prey1, aes(x = Mountain, y = Order, size = Abundance), alpha = 0.7, color = "brown") + 
  geom_count(data = prey1, aes(x = Urban, y = Order, size = Abundance), alpha = 0.7, color = "#8f6e98") + 
  geom_count(data = prey1, aes(x = Coastal, y = Order, size = Abundance), alpha = 0.7, color = "pink") + 
  geom_count(data = prey1, aes(x = Unknown, y = Order, size = Abundance), alpha = 0.7, color = "black") + 
  facet_wrap(~bat.sp) + theme_bw() + 
  xlab("") + ylab("") + 
  theme(axis.text.x=element_blank(), axis.ticks.x= element_blank()) +
  theme(text = element_text(size = 20)) + xlim(c(1,8)) +
  theme(legend.position = "none") 
  
p # 

# ## How well do these two compare? - many gaps, lets look at the most prevalent prey items. 
# test <- left_join(taxasimple, swbugsh3, by = "Species") # 1887 obs 
# test1 <- test %>% distinct(Species, .keep_all = TRUE) 
# test2 <- unique(test1)
# 
# # 40 most prevalent prey items for both bats combined 
# prey1 <- prevalentpreyall %>% distinct(Species, .keep_all = TRUE) 
# prey1$Species # Not actually that useful for comparing between the two species... 
# 
# ## Better to split it up and then re-merge
# preyB <- prevalentpreyMBRA %>% distinct(Species, .keep_all = TRUE) 
# preyB$BatSpecies <- "M. brandtii" # select the top 30 prey items for MBRA
# preyB$Species
# # 162 distinct prey items - but not the same ones as MMYS! - but there are many with zero abundance and prevalence... 
# 
# preyM <- prevalentpreyMMYS %>% distinct(Species, .keep_all = TRUE) 
# preyM$BatSpecies <- "M. mystacinus"
# preyM$Species
# # 162 distinct prey items - but not the same as MBRA! - but there are many with zero abundance and prevalence... 
# 
# # How many actual prey items for each bat species? 
# preycheckB <- subset(preyB, preyB$Prevalence != 0 & preyB$TotalAbundance != 0) 
# #67 prey items
# 
# preycheckM <- subset(preyM, preyM$Prevalence != 0 & preyM$TotalAbundance != 0) 
# # 103 prey items 
# 
# ## For just comparing all prey items of the two bats, this object should be saved and stored 
# 
# preycheckBM <- full_join(preycheckB, preycheckM) %>% arrange(Order)
# 



```


### Plotting abundance, prevalence, relative abundance across prey orders 

```{r}
## Just differences in number of counts 
  ggplot() + 
  geom_count(data = prey1, aes(x = BatSpecies, y = Family, color = Order), alpha = 0.5) + 
  scale_size(range = c(5, 15)) +
  theme_bw() + xlab("") + ylab("") + 
  theme(text = element_text(size = 20)) 

# Total Abundance 
  ggplot() + 
  geom_count(data = prey1, aes(x = BatSpecies, y = Family, size = TotalAbundance, color = Order), alpha = 0.5) + 
  scale_size(range = c(5, 12)) +
  theme_bw() + xlab("") + ylab("") +
  theme(text = element_text(size = 20)) +
  geom_jitter()

  
  # Prevalence
  ggplot() + 
  geom_count(data = prey1, aes(x = BatSpecies, y = Family, size = Prevalence, color = Order), alpha = 0.5) + 
  scale_size(range = c(5, 12)) +
  theme_bw() + xlab("") + ylab("") +
  theme(text = element_text(size = 20)) +
  geom_jitter()
  
################################################# 
### Use this ###
################################################# 
  
  ## FamilyOrders
  checkthis <- prey1 %>% dplyr::select(Family, Order) %>% 
    distinct() %>% 
    mutate(Family = factor(Family)) %>% 
    mutate(Order = factor(Order))
  # 53 levels 

  ## Aggregate to Family level then use RRA - 
prey2 <- prey1 %>% group_by(BatSpecies, Sample, Family, Order) %>% 
    dplyr::summarize(FamAbun = sum(TotalAbundance)) 
# NAs introduced where there should just be an Order for Family 
prey2$Family <- as.character(prey2$Family)
prey2 <- prey2 %>% mutate(Family = replace_na(Family, "FixThis"))
prey2 <- prey2 %>% mutate(Family = case_when(
  Family == "FixThis" ~ paste0(as.character(Order), " Order"), 
  TRUE ~ as.character(Family))) %>% mutate(Family = factor(Family))  %>% arrange(Order)

  checkthis <- prey2 %>% dplyr::select(Family, Order) %>% distinct() %>% arrange(Order) 
  listcheck <- as.character(checkthis$Family) %>% unique() # list of 53, good 
  prey2$Family <- factor(prey2$Family, levels = listcheck) 
  prey2$Order <- factor(prey2$Order) # Create a nice pallete with 10 distinct colors 
  
  met.brewer("Hiroshige", n = 10)
  met.brewer("OKeeffe1", n = 10)
  met.brewer("Paquin", n = 10)
  met.brewer("Hokusai2", n = 10)
  
  summary(prey2)
  
 #    BatSpecies           Sample                     Family             Order         FamAbun     
 # Length:1909        Length:1909        Limoniidae    :  47   Diptera    :877   Min.   :   18  
 # Class :character   Class :character   Pediciidae    :  47   Lepidoptera:579   1st Qu.:  153  
 # Mode  :character   Mode  :character   Tachinidae    :  47   Trichoptera:141   Median :  617  
 #                                       Tipulidae     :  47   Neuroptera : 78   Mean   : 3160  
 #                                       Chironomidae  :  47   Hymenoptera: 63   3rd Qu.: 2799  
 #                                       Bolitophilidae:  47   Psocodea   : 62   Max.   :50868  
 #                                       (Other)       :1627   (Other)    :109                 
 #  
  
  
## THIS goes into the paper   
# Now the families are arranged by Order!!  NICE

p <-  ggplot() + 
  geom_count(data = prey2, aes(x = BatSpecies, y = Family, size = FamAbun, color = Order), alpha = 0.9,) + 
  scale_size_continuous(range = c(5, 12)) +
  scale_color_manual(values = met.brewer("Signac", n = 10)) +
  xlab("") + ylab("") +
  theme(
  panel.background = element_rect(fill = "white",
                                colour = "white",
                                size = 0.5, linetype = "solid"),
  panel.grid.major = element_line(size = 0.5, linetype = 'solid',
                                colour = "#7F7F7F"), 
  panel.grid.minor = element_line(size = 0.25, linetype = 'solid',
                                colour = "#7F7F7F")) + 
  theme(text = element_text(size = 20)) +
  theme(legend.title = element_blank()) +
  geom_jitter()
p + guides(size = FALSE)
# Range of dot sizes between 1000 and 50000 

###### Proportion of which prey is represented in between the two specie - Relative Abundance
##### No difference if this is relative or total abundance or even prevalence because it is a proportion 
propbar <- ggplot(prey1) +
  geom_bar(aes(x=BatSpecies, y = Prevalence, fill = Order), stat = "identity", position = "fill") + 
  scale_y_continuous(labels = scales::percent) +
  scale_fill_manual(values = met.brewer("Signac", n = 10)) +
  xlab("") + ylab("") +
  theme(
  panel.background = element_rect(fill = "white",
                                colour = "white",
                                size = 0.5, linetype = "solid"),
  panel.grid.major = element_line(size = 0.5, linetype = 'solid',
                                colour = "#7F7F7F"), 
  panel.grid.minor = element_line(size = 0.25, linetype = 'solid',
                                colour = "#7F7F7F")) + 
  theme(text = element_text(size = 25)) +
  theme(legend.position = "none") 
propbar

### Bipartite

## Make a bipatriate chart 

bipart <- ggplot(prey1, aes(x = BatSpecies, y = Order, fill = TotalAbundance)) +
  geom_tile() +
  scale_fill_gradient(low = "gray", high = "black") + 
  labs(title = "",
       x = "",
       y = "") +  
  theme(text = element_text(size = 25)) +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = element_blank(), 
        axis.line = element_line(colour = "black")) +
  theme(legend.position = "none") 
bipart


### Put those together to make one figure (done separately in publisher)
```


 
 
 
```{r, echo = TRUE ,results = 'hide',fig.keep = 'all'}



## Remember. This changes each time you run it. Eventually, should just save a csv file of one output for this and reuse it in a separate analysis probably. 
bbord <- ordinate(batsbugsy, "NMDS", distance="bray", k = 4)
bbord 
# NMDS started throwing errros (
#Warning message:
# In postMDS(out$points, dis, plot = max(0, plot - 1), ...) :
#   skipping half-change scaling: too few points below threshold)  
# so I switched to PCoA 

#         

# envfit() from vegan to stack on top of ordination. Typically best for continuous variables, also accepts factor variables. This is an alternative to ANOVA. Not a real statistcal test. 

O12<- plot_ordination(batsbugsy, bbord, color = "bat.sp", axes = c(1,2)) + 
  stat_conf_ellipse(aes(fill=bat.sp ),alpha=0.25,geom="polygon",level=0.95) + 
  theme_bw() + 
  ggtitle("NMDS1 and 2 with Bray-Curtis difference index ")
O12


O13<- plot_ordination(batsbugsy, bbord, color = "bat.sp", axes = c(1,3)) + 
  stat_conf_ellipse(aes(fill=bat.sp ),alpha=0.25,geom="polygon",level=0.95) + 
  theme_bw()+ 
  ggtitle("NMDS1 and 3 with Bray-Curtis difference index ")
O13

O14<- plot_ordination(batsbugsy, bbord, color = "bat.sp", axes = c(1,4)) + 
  stat_conf_ellipse(aes(fill=bat.sp ),alpha=0.25,geom="polygon",level=0.95) + 
  theme_bw() + 
  ggtitle("NMDS1 and 4 with Bray-Curtis difference index ")
O14 


O23 <- plot_ordination(batsbugsy, bbord, color = "bat.sp", axes = c(2,3)) + 
  stat_conf_ellipse(aes(fill=bat.sp ),alpha=0.25,geom="polygon",level=0.95) + 
  theme_bw() + 
  ggtitle("NMDS 2 and 3 with Bray-Curtis difference index ")
O23 

O24 <- plot_ordination(batsbugsy, bbord, color = "bat.sp", axes = c(2,4)) + 
  stat_conf_ellipse(aes(fill=bat.sp ),alpha=0.25,geom="polygon",level=0.95) + 
  theme_bw() + 
  ggtitle("NMDS 2 and 4 with Bray-Curtis difference index ")
O24

O34 <- plot_ordination(batsbugsy, bbord, color = "bat.sp", axes = c(3,4)) + 
  stat_conf_ellipse(aes(fill=bat.sp ),alpha=0.25,geom="polygon",level=0.95) + 
  theme_bw() + 
  ggtitle("NMDS 3 and 4 with Bray-Curtis difference index ")
O34


### Start here tomorrow 

#Create a dataframe with NMDS scores
# ordscore <- scores(bbord, choices=c(1,2,3,4,5), tidy = TRUE)
# ordscore <- as.data.frame(ordscore)
# ordscore <- cbind(rownames(ordscore), data.frame(ordscore, row.names=NULL))
# names(ordscore)[1] <- "Sample"
#1934 obs of 8 vars 
 
```

# Mixed linear model 

https://support.minitab.com/en-us/minitab/18/help-and-how-to/modeling-statistics/anova/how-to/mixed-effects-model/interpret-the-results/key-results/#step-2-determine-whether-the-fixed-effect-terms-significantly-affect-the-response


```{r, echo = TRUE ,results = 'hide',fig.keep = 'all'}
bbdf <- ps_melt(batsbugsy) # convert to a df 
bbdf1 <- left_join(bbdf, ordscore) # add the NMDS scores

## use this data frame for modeling further: 

#write.csv(bbdf1, file=file.path(output_today,"mbra_mmys_NMDS_Scores_diet.csv"))

### use NMDS1 as the response variable 
library(nlme)

# simplest model 
m1 <- lme(NMDS1~bat.sp, random=~1|label, data=bbdf1)
summary(m1)

# +year
m2 <- lme(NMDS1~bat.sp + year, random=~1|label,data=bbdf1)
summary(m2)

# species*year interaction 
m3 <- lme(NMDS1~bat.sp + bat.sp*year, random=~1|label, data=bbdf1)
summary(m3)


## Ask Katrine what the best way is to compare model selection for these kinds of NMDS models... 

# anova(m1, m2, m3, m4)
# 
# 
# anova(m1_int, m2_int, m3_int, m4_int)


#Quick visualization of difference between the two species:

par(mfrow=c(2, 1))
plot(allEffects(m1), type = "response")
plot(allEffects(m2), type = "response")
plot(allEffects(m3), type = "response")



# https://rdrr.io/cran/sjPlot/f/vignettes/plot_marginal_effects.Rmd

#plotting marginal effects
#plot_model(mod1, type = "pred", terms = "bat.sp")
plot_model(m3, type = "eff", terms = "bat.sp")

