
---
title: "Nittedal Diet Analyses"
author: "Reed McKay"
date: "4 9 2021"
output: html_document
---



### Previous versions of the script: 

title: "IBBM Diet Analyses"
author: "Reed McKay"
date: "4 5 2021"
output: html_document


title: "Filter/Normal_NMDS_lmer_batsandbugs"
author: "Reed April McKay"
date: "16 2 2021"
output: html_document


### Prepare work space and load packages 

```{r setup, include=FALSE}
library(rlang)
library(cowplot)
library(tidyverse) #
library(readxl) #
library(iNEXT) # 
library(vegan) #
library(reshape2) 
library(ggpubr) # 
library(cowplot) # 
library(devtools) #
library(lme4) 
library(phyloseq)
library(plyr)
library(Rcpp)
library(vctrs)
library(metagMisc)
library(DHARMa)
library(sjPlot)
library(effects)
library(qiime2R)
library(BiocManager)
library(patchwork)
library(microViz)
library(beepr)
library(microViz)
library(mixOmics)
library(ade4)
library(nlme)
library(pals)
library(RColorBrewer)

# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")

# The following initializes usage of Bioc devel
# BiocManager::install(version='devel')

library(devtools) # Load the devtools package
# install_github("microbiome/microbiome") # Install the package
# devtools::install_github("microsud/microbiomeutilities")
library(microbiome)
library(microbiomeutilities)
library(MASS)
#devtools::install_github("jbisanz/qiime2R")
library(qiime2R)

# Installing from CRAN
#install.packages("sorvi")
library(sorvi)
# Installing from Github
#install_github("antagomir/netresponse")
library(netresponse)

getwd()
# C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/Documents/1. PhD_Main/GitHub_link/Nittedal/Mbra-vs-Mmys-diet-"

output <- "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs"

file.name <- "Nittedal_modeling pt1"

todays_date <- Sys.Date()
 
dir.name <- str_c(output,"/", file.name, "_", todays_date)
dir.name
 
output_today <- dir.name
output_today

dir.create(output_today)
output_today

```

### Import data, create phyloseq object, convert OTU table to presence/absence (optional)

```{r, echo = TRUE ,results = 'hide',fig.keep = 'all'}

phylo <- qza_to_phyloseq(
  features= "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/FromFranz_07.2023/R-scrips/sPLS-DA/bat-feat-tab-clean.qza",  
  # This file was sent by Franz on 06.09.2023        
   #it makes sense to use a filtered table, as sPLS-DA can't handle rare features very well
  taxonomy= "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/FromFranz_07.2023/R-scrips/sPLS-DA/bat-classified-repr-seq.qza", 
  metadata= "C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/FromFranz_07.2023/R-scrips/sPLS-DA/bat-metadata_clean.tsv")

phylo #shows what our physeq object contains
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 3629 taxa and 64 samples ]
# sample_data() Sample Data:       [ 64 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 3629 taxa by 7 taxonomic ranks ]



sort(colSums(otu_table(phylo))) # 4000 reads per sample 

#we remove all features that are unassigned on phylum level
phylos <- subset_taxa(phylo, Phylum != "NA")

phylos   #shows what our physeq object contains
#phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 3629 taxa and 64 samples ]
# sample_data() Sample Data:       [ 64 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 3629 taxa by 7 taxonomic ranks ]
#now we extract the data we need in the exact format that we need

#we remove all features that are unassigned on genus level
phyloseq <- subset_taxa(phylo, Genus != "NA")

phyloseq   #shows what our physeq object contains
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 2536 taxa and 64 samples ]
# sample_data() Sample Data:       [ 64 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 2536 taxa by 7 taxonomic ranks ]

# 1093 taxa lost 


#we remove all features that are unassigned on family level
phyloseq1 <- subset_taxa(phylo, Family != "NA")

phyloseq1   #shows what our physeq object contains
# phyloseq1   #shows what our physeq object contains
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 2665 taxa and 64 samples ]
# sample_data() Sample Data:       [ 64 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 2665 taxa by 7 taxonomic ranks ]
#3629 - 2665 =  964


#we remove all features that are unassigned on genus level
phyloseq2 <- subset_taxa(phylo, Order != "NA")

phyloseq2   #shows what our physeq object contains
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 3629 taxa and 64 samples ]
# sample_data() Sample Data:       [ 64 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 3629 taxa by 7 taxonomic ranks ]*

### Franz has suggested NOT filtering out taxa with NA values at this stage.


batsbugs <- phylo %>%  ps_filter(bat.sp %in% c("MBRA", "MMYS"))
batsbugs  
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 2616 taxa and 47 samples ]
# sample_data() Sample Data:       [ 47 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 2616 taxa by 7 taxonomic ranks ]
#   
#Inspect the Phyloseq object:
  
  sample_names(batsbugs)
  rank_names(batsbugs)
  sample_variables(batsbugs)
  
sort(colSums(otu_table(batsbugs)))
quantile(colSums(otu_table(batsbugs)))
#   0%  25%  50%  75% 100% 
# 4000 4000 4000 4000 4000 
```


### Filter and Normalize
```{r, echo = TRUE ,results = 'hide',fig.keep = 'all'}

#Filter1 = batbugs1 

#Remove OTUs with less than 10 sequence reads 
batsbugs1 <- prune_taxa(taxa_sums(batsbugs) >= 10, batsbugs) 
batsbugs
batsbugs1
# 2616 - 1887 # 729 taxa lost 

ntaxa(batsbugs1)
quantile(colSums(otu_table(batsbugs1)))
#     0%    25%    50%    75%   100% 
# 3857.0 3934.0 3950.0 3960.5 4000.0 

sort(colSums(otu_table(batsbugs1)))

# Remove samples with unusually small amounts of reads  (less tan 600)
# Removed four samples:     13B   3B  49B   5B  11B
# (recommendation from Lisa Lunde)

batsbugs2<-prune_samples(colSums(otu_table(batsbugs1))>=600,batsbugs1) 
batsbugs2 # none lost 
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 1887 taxa and 47 samples ]
# sample_data() Sample Data:       [ 47 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 1887 taxa by 7 taxonomic ranks ]

#removed 5 samples, check that this worked: 
sort(colSums(otu_table(batsbugs2)))
quantile(colSums(otu_table(batsbugs2)))
#     0%    25%    50%    75%   100% 
# 3857.0 3934.0 3950.0 3960.5 4000.0 


# Compare abundance across phyloseq objects 

p0 <- plot_bar(batsbugs, x = "bat.sp", fill = "Order") +
                geom_bar(stat="identity") + theme_bw() + ggtitle("Raw")
p0 # raw data 


p1 <- plot_bar(batsbugs2, x = "bat.sp", fill = "Order") +
                geom_bar(stat="identity") + theme_bw() + ggtitle("Samples > 600 reads")
p1 # batsbugs 2 and removed samples with less than 600 reads 

```

Relative abundance representation
Pooled samples
Normalized ad filtered data 
Top Orders 

https://mibwurrepo.github.io/Microbial-bioinformatics-introductory-course-Material-2018/composition-plots.html

```{r}

## MICROBIOMEUTILITIES IS THE NEW ONE TO USE! 

bb2.fam <- aggregate_top_taxa2(batsbugs2, "Family", top = 70)

plot.composition.CountAbun <- plot_composition(bb2.fam) + 
  theme_bw() +ggtitle("Top 70 famillies")
  
plot.composition.CountAbun


#Make relative abundance 

bb.fam.rel <- microbiome::transform(bb2.fam, "compositional")

plot_bar(bb.fam.rel, x="bat.sp", fill = "Family")+ xlab(NULL) + geom_bar(stat="identity") 

plot_composition(bb.fam.rel, fill = "family", average_by = "bat.sp") + 
  labs( x = "Bat Species", y = "Relative abundance", title = "Relative Abundance of Prey Item by Family") 


```

Now for order

```{r}

bb2.order <- aggregate_top_taxa2(batsbugs2, "Order", top = 45) 

plot.composition.CountAbun <- plot_composition(bb2.order) + 
  theme_bw() + ggtitle("Top 45 orders") + facet_wrap(~ bb2.order@sam_data$bat.sp)
  
plot.composition.CountAbun


#Make relative abundance 


bb.order.rel <- transform(bb2.order, "compositional")

plot_bar(bb.order.rel, x="bat.sp", fill = "Order")+ xlab(NULL) + 
  geom_bar(stat="identity") + 
  labs( y = "Relative abundance",
  title = "Relative Abundance of Prey Item by Order")  



plot_composition(bb.order.rel, fill = "Order", average_by = "bat.sp") + 
  labs( x = "Bat Species", y = "Relative abundance", 
        title = "Relative Abundance of Prey Item by Order") +
  theme_bw()

plot.composition.relAbun <- plot_composition(bb.order.rel, sample.sort = "bat.sp") + facet_wrap(~ bb.order.rel@sam_data$bat.sp)
print(plot.composition.relAbun)

```



## Summarizing and visualizing RAW data between the two bat species

*You can more easily see the differences* in this Google Sheet where I copied these summaries: https://docs.google.com/spreadsheets/d/11VWNq5zbMou82J_H8RsgL3eJWfj56kPKXcp8vh6KHcA/edit#gid=0 

*Mostly, there aren't any difference except in abundance of prey items for these top 10 OTUs* 

How many prey items for MBRA and MMYS in the filtered data? 

```{r, echo = TRUE ,results = 'hide',fig.keep = 'all'}

# Taxa with positive sum across samples
bb_prune <- prune_taxa(taxa_sums(batsbugs2) > 0, batsbugs2)
ntaxa(bb_prune)
# 1887
ntaxa(batsbugs2)
# 1887


#Subset the raw data by bat species and summarize OTU abundance

# MBRA 

MBRA <- subset_samples(batsbugs2, bat.sp == "MBRA")
MBRA
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 1887 taxa and 16 samples ]
# sample_data() Sample Data:       [ 16 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 1887 taxa by 7 taxonomic ranks ]

# B_prune = prune_taxa(taxa_sums(MBRA) > 0, MBRA)
# ntaxa(B_prune)
# B_prune
# # phyloseq-class experiment-level object
# # otu_table()   OTU Table:         [ 832 taxa and 16 samples ]
# # sample_data() Sample Data:       [ 16 samples by 9 sample variables ]
# # tax_table()   Taxonomy Table:    [ 832 taxa by 7 taxonomic ranks ]

# For future potential color pallettes

# Diptera", 
 "#126782"

#"Hemiptera", 
  "#A5E1F3"
  
#"Hymenoptera", 
  "#023047"

#"Lepidoptera", 
  "#425326"
   
#"Neuroptera", 
   "#817425"

#"Trichoptera"
   "#FB8500"

plot_taxa_prevalence(MBRA, "Order") 
quantile(colSums(otu_table(MBRA)))
sort(colSums(otu_table(MBRA)))

######################################################

#MMYS
MMYS <- subset_samples(batsbugs2, bat.sp == "MMYS")
MMYS
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 1887 taxa and 31 samples ]
# sample_data() Sample Data:       [ 31 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 1887 taxa by 7 taxonomic ranks ]

# M_prune = prune_taxa(taxa_sums(MMYS) > 0, MMYS)
# ntaxa(M_prune)
# M_prune
# phyloseq-class experiment-level object
# otu_table()   OTU Table:         [ 1249 taxa and 31 samples ]
# sample_data() Sample Data:       [ 31 samples by 9 sample variables ]
# tax_table()   Taxonomy Table:    [ 1249 taxa by 7 taxonomic ranks ]

#Mtaxaorder <- 
 
# Diptera", 
  "#126782"
  
#"Hymenoptera", 
  "#023047"

#"Lepidoptera", 
  "#425326"
   
#"Neuroptera", 
   "#817425"

#"Orthoptera",
   "#FFB703"

#"Psocodea", 
   "#FD9E02"

#"Trichoptera"
   "#FB8500"

plot_taxa_prevalence(MMYS, "Order") 
quantile(colSums(otu_table(MMYS)))
sort(colSums(otu_table(MMYS)))

######################################################


# Visualize and compare - ORders 

pB <- plot_bar(MBRA, x = "bat.sp", fill = "Order") + 
  xlab(NULL) + geom_bar(stat="identity") + theme_bw() 
pB


pM <- plot_bar(MMYS, x = "bat.sp", fill = "Order") +
  xlab(NULL) + geom_bar(stat="identity") + theme_bw() 
pM

cowplot::plot_grid(pB, pM) + 
  scale_colour_brewer(palette="Set1") 

# Visualize and compare - families 
tax_glom(MBRA, "Family") # 49 different families for MBRA
tax_glom(MMYS, "Family") # 49 different families for MMYS

tax_glom(MBRA, "Genus") # 101 different families for MBRA
tax_glom(MMYS, "Genus") # 101 different families for MMYS

# colourCount = length(unique(mtcars$hp))
# getPalette = colorRampPalette(brewer.pal(9, "Set1"))
#  
# ggplot(mtcars) + 
#   geom_histogram(aes(factor(hp)), fill=getPalette(colourCount)) + 
#   theme(legend.position="right")


### Other cool plotting options for later if I want to make something like this for sharing/publishing 
# colorcount1 <- 30 
# colorcount2 <- 35
# 
# pB <- plot_bar(B_prune, x = "bat.sp", fill = "Family") + 
#   xlab(NULL) + geom_bar(stat="identity") + theme_bw() + 
#   scale_fill_manual(values = colorRampPalette(brewer.pal(11, "Spectral"))(colorcount1))
# pB
# 
# pM <- plot_bar(M_prune, x = "bat.sp", fill = "Family") +
#   xlab(NULL) + geom_bar(stat="identity") + theme_bw() + 
#   scale_fill_manual(values = colorRampPalette(brewer.pal(11, "Spectral"))(colorcount2))
# pM
# 
# cowplot::plot_grid(pB, pM)


######################################################

# Prevalence of the top 10 most abundant ASVs for both bat species from raw data 
### Now there are some NAs in the top prey items - we may able to figure out what these are later by blasting the data through NCBI

prevelancedf = apply(X = otu_table(bb_prune),
                 MARGIN = 1,
                 FUN = function(x){sum(x > 0)})
# Add taxonomy and total read counts to this data.frame
prevelancedf = data.frame(Prevalence = prevelancedf,
                      TotalAbundance = taxa_sums(bb_prune),
                      tax_table(bb_prune))
prevelancedf[1:20,]


######################################################

# Prevalence of the top 10 most abundant OTUs for MBRA from raw data

prevelancedf = apply(X = otu_table(MBRA),
                 MARGIN = 1,
                 FUN = function(x){sum(x > 0)})
# Add taxonomy and total read counts to this data.frame
prevelancedf = data.frame(Prevalence = prevelancedf,
                      TotalAbundance = taxa_sums(MBRA),
                      tax_table(MBRA))
prevelancedf[1:20,]

######################################################

# Prevalence of the top 10 most abundant OTUs for MMYS from raw data

prevelancedf = apply(X = otu_table(MMYS),
                 MARGIN = 1,
                 FUN = function(x){sum(x > 0)})
# Add taxonomy and total read counts to this data.frame
prevelancedf = data.frame(Prevalence = prevelancedf,
                      TotalAbundance = taxa_sums(MMYS),
                      tax_table(MMYS))
prevelancedf[1:20,]



########################################################

## Find the NAs 

test <- left_join(bb_prune, seqs)

# Convert Phyloseq object to dataframe
bbdf = psmelt(bb_prune)
dim(bbdf)
#88689    19


summary(bbdf)
names(bbdf)
#new_DF <- DF[is.na(DF$Var),]
bbdf1<- bbdf[is.na(bbdf$Family),]
### START HERE! 
# 19317 obs of NA on at least the family level. 

# But how many unique ASVs? 

bbdf1$OTU <- as.factor(bbdf1$OTU)
summary(bbdf1)
        

# remove NAs
bbdf <- bbdf[complete.cases(bbdf),]
#49800 obs of 19 vars
summary(bbdf)
str(bbdf)

# Add the NMDS scores 
bbdf1 <- left_join(bbdf, ordscore, by = "Sample")
names(bbdf1)
summary(bbdf1) # 411 unique levels of unknown ASVs 

na_asv <- bbdf1 %>% distinct(OTU, .keep_all=T) 
write.csv(na_asv, file=file.path(output_today,"unknown_asvs.csv")) 
output_today

## Check which sequences are related to these features
seqs <- read_qza("C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Analysis archive/bat_repr-seq.qza")
str(seqs)
seqs$OTU <- seqs$uuid
dim(seqs)
dim(na_asv)

na_seq <- cross_join(na_asv, seq1, copy = TRUE)
dim(na_seq)
# 2065686      20

na_seq1 <- na_seq %>% distinct(OTU, .keep_all = TRUE) 
# write.csv(na_seq1, file=file.path(output_today,"unknown_asvs_with_sequences.csv")) 

na_seq2 <- na_seq1 %>% dplyr::select(OTU, x)
# write.csv(na_seq2, file=file.path(output_today,"unknown_asvs_with_sequences_only.csv")) 

fasta_prep <- na_seq2 %>% rename(seqs = x, Names = OTU)

seqlist <- as.list(fasta_prep$seqs)
seqnames <- as.list(fasta_prep$Names)

```

## NCBI BLAST for unknown features  

vignette: https://rsh249.github.io/bioinformatics/rBlast.html#get_data
https://github.com/mhahsler/rBLAST/blob/master/README.md


```{r}
library(seqinr)
library(taxonomizr)
#install.packages("devtools")
#devtools::install_github("mhahsler/rBLAST")
library(rBLAST)

fname <- tempfile(pattern = "UnknownPrey", tmpdir = output_today, fileext = ".fasta")
write.fasta(seqlist, names = seqnames, as.string = FALSE, file.out =fname)
fname

b1 <- blast("C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs/Nittedal_modeling pt1_2023-09-11a/UnknownPrey.fasta")

# Example data 
# seq <- readRNAStringSet(system.file("C:/Users/apmc/OneDrive - Norwegian University of Life Sciences/1. Nittedal 2018-2020/Nittedal_Main/Analyses/Outputs/Nittedal_modeling pt1_2023-09-11a/unknown_asvs_with_sequences_only.csv"), package = "rBLAST") 


```



 
```{r, echo = TRUE ,results = 'hide',fig.keep = 'all'}

bbord <- ordinate(bb_prune, "NMDS", distance="bray", k = 4)
bbord


## Remember. This changes each time you run it. Eventually, should just save a csv file of one output for this and reuse it in a separate analysis probably. 


# envfit() from vegan to stack on top of ordination. Typically best for continuous variables, also accepts factor variables. This is an alternative to ANOVA. Not a real statistcal test. 

O12<- plot_ordination(bb_prune, bbord, color = "bat.sp", axes = c(1,2)) + 
  stat_conf_ellipse(aes(fill=bat.sp ),alpha=0.25,geom="polygon",level=0.95) + 
  theme_bw() + 
  ggtitle("NMDS1 and 2 with Bray-Curtis difference index ")
O12


O13<- plot_ordination(bb_prune, bbord, color = "bat.sp", axes = c(1,3)) + 
  stat_conf_ellipse(aes(fill=bat.sp ),alpha=0.25,geom="polygon",level=0.95) + 
  theme_bw()+ 
  ggtitle("NMDS1 and 3 with Bray-Curtis difference index ")
O13

O14<- plot_ordination(bb_prune, bbord, color = "bat.sp", axes = c(1,4)) + 
  stat_conf_ellipse(aes(fill=bat.sp ),alpha=0.25,geom="polygon",level=0.95) + 
  theme_bw() + 
  ggtitle("NMDS1 and 4 with Bray-Curtis difference index ")
O14 


O23 <- plot_ordination(bb_prune, bbord, color = "bat.sp", axes = c(2,3)) + 
  stat_conf_ellipse(aes(fill=bat.sp ),alpha=0.25,geom="polygon",level=0.95) + 
  theme_bw() + 
  ggtitle("NMDS 2 and 3 with Bray-Curtis difference index ")
O23 

O24 <- plot_ordination(batsbugs2, bbord, color = "bat.sp", axes = c(2,4)) + 
  stat_conf_ellipse(aes(fill=bat.sp ),alpha=0.25,geom="polygon",level=0.95) + 
  theme_bw() + 
  ggtitle("NMDS 2 and 4 with Bray-Curtis difference index ")
O24

O34 <- plot_ordination(batsbugs2, bbord, color = "bat.sp", axes = c(3,4)) + 
  stat_conf_ellipse(aes(fill=bat.sp ),alpha=0.25,geom="polygon",level=0.95) + 
  theme_bw() + 
  ggtitle("NMDS 3 and 4 with Bray-Curtis difference index ")
O34


#Create a dataframe with NMDS scores
ordscore <- scores(bbord, choices=c(1,2,3,4), tidy = TRUE)
ordscore <- as.data.frame(ordscore)
ordscore <- cbind(rownames(ordscore), data.frame(ordscore, row.names=NULL))
names(ordscore)[1] <- "Sample"
#1934 obs of 7 vars 
 
```

# Mixed linear model 

https://support.minitab.com/en-us/minitab/18/help-and-how-to/modeling-statistics/anova/how-to/mixed-effects-model/interpret-the-results/key-results/#step-2-determine-whether-the-fixed-effect-terms-significantly-affect-the-response


```{r, echo = TRUE ,results = 'hide',fig.keep = 'all'}


### START HERE 
library(nlme)
m1 <- lme(NMDS1~bat.sp + year, random=~1|label, data=bbdf1)
summary(m1)


m2 <- lme(NMDS2~bat.sp + year, random=~1|label,data=bbdf1)
summary(m2)


m3 <- lme(NMDS3~bat.sp + year, random=~1|label,data=bbdf1)
summary(m3)


m4 <- lme(NMDS4~bat.sp + year, random=~1|label,data=bbdf1)
summary(m4)

################ Test with year*bat.sp interaction 
m1_int <- lme(NMDS1~bat.sp + bat.sp*year, random=~1|label,data=bbdf1)
summary(m1_int)

m2_int <- lme(NMDS2~bat.sp + bat.sp*year, random=~1|label,data=bbdf1)
summary(m2_int)

m3_int <- lme(NMDS3~bat.sp + bat.sp*year, random=~1|label,data=bbdf1)
summary(m3_int)

m4_int <- lme(NMDS4~bat.sp + bat.sp*year, random=~1|label,data=bbdf1)
summary(m4_int)


## Ask Katrine what the best way is to compare model selection for these kinds of NMDS models... 

# anova(m1, m2, m3, m4)
# 
# 
# anova(m1_int, m2_int, m3_int, m4_int)


#Quick visualization of difference between the two species:

par(mfrow=c(2, 1))
plot(allEffects(m1), type = "response")
plot(allEffects(m2), type = "response")

# https://rdrr.io/cran/sjPlot/f/vignettes/plot_marginal_effects.Rmd

#plotting marginal effects
#plot_model(mod1, type = "pred", terms = "bat.sp")
plot_model(m1, type = "eff", terms = "bat.sp")

